<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast Multivariate Analyses of Big Genomic Data • smartsnp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast Multivariate Analyses of Big Genomic Data">
<meta property="og:description" content="Fast computation of multivariate analyses of small (10s to 100s markers) to big (1000s to 100000s) genotype data. Runs Principal Component Analysis allowing for centering, z-score standardization and scaling for genetic drift, projection of ancient samples to modern genetic space and multivariate tests for differences in group location (Permutation-Based Multivariate Analysis of Variance) and dispersion (Permutation-Based Multivariate Analysis of Dispersion).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="smartsnp" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#smartsnp" class="anchor"></a>smartsnp</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The package <em>smartsnp</em> runs fast and user-friendly computation of Principal Component Analysis (PCA) on single-nucleotide-polymorphism (SNP) data suitable for ancient, low-coverage and modern DNA. The package combines SNP scaling for genetic drift and projection of ancient samples onto a modern genetic PCA space (currently available only in Unix environment in the field-standard software EIGENSOFT) with permutation-based multivariate tests for population differences in genetic diversity (both location and dispersion). The package comprises three functions that run each analysis individually (<em>smart_pca</em>, <em>smart_permanova</em>, <em>smart_permdisp</em>), and a wrapper function (<em>smart_mva</em>) that runs any combination of the three standalone functions.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of smartsnp from <a href="https://CRAN.R-project.org">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"smartsnp"</span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is an example of how to run PCA, PERMANOVA and PERMDISP controlling for genetic drift for the package’s dataset <em>dataSNP</em> including 10000 simulated SNPs in 100 samples (80 = modern, 20 = ancient).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#1/ Load package and label samples</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">smartsnp</span><span class="op">)</span>
<span class="co"># Path to example genotype matrix "dataSNP"</span>
<span class="va">pathToGenoFile</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dataSNP"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="co">#assign 50 samples to each of two groups</span>
<span class="va">my_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span>
<span class="co">#assign samples 1st to 10th per group to ancient</span>
<span class="va">my_ancient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">51</span><span class="op">:</span><span class="fl">60</span><span class="op">)</span>

<span class="co">#2/ Run PCA with truncated SVD (PCA 1 x PCA 2 axes) and assign results to object pcaR</span>
<span class="va">pcaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/smart_pca.html">smart_pca</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToGenoFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, sample_project <span class="op">=</span> <span class="va">my_ancient</span><span class="op">)</span>
<span class="co">#assign statistical results to objects pcaR_eigen, pcaR_load and pcaR_coord</span>
<span class="va">pcaR_eigen</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.eigenvalues</span>; <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pcaR_eigen</span><span class="op">)</span> <span class="co"># extract eigenvalues</span>
<span class="co">#&gt; [1] 3 2</span>
<span class="va">pcaR_load</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.snp_loadings</span>; <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pcaR_load</span><span class="op">)</span> <span class="co"># extract principal coefficients (SNP loadings)</span>
<span class="co">#&gt; [1] 4532    2</span>
<span class="va">pcaR_coord</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.sample_coordinates</span>; <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pcaR_coord</span><span class="op">)</span> <span class="co"># extract principal components (sample position in PCA space)</span>
<span class="co">#&gt; [1] 100   4</span>

<span class="co">#3/ Run PERMANOVA test (group location in PCA1 x PCA2 space after excluding ancient samples) and assign results to object permanovaR</span>
<span class="va">permanovaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/smart_permanova.html">smart_permanova</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToGenoFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, target_space <span class="op">=</span> <span class="st">"pca"</span>, sample_remove <span class="op">=</span> <span class="va">my_ancient</span><span class="op">)</span>
<span class="co">#assign sample summary to object permP</span>
<span class="va">permP</span> <span class="op">&lt;-</span> <span class="va">permanovaR</span><span class="op">$</span><span class="va">permanova.samples</span>
<span class="co">#show PERMANOVA table</span>
<span class="va">permanovaR</span><span class="op">$</span><span class="va">permanova.global_test</span>
<span class="co">#&gt;           Df SumsOfSqs MeanSqs F.Model      R2 Pr(&gt;F)</span>
<span class="co">#&gt; group      1     175.8  175.82 0.45613 0.00581 0.6459</span>
<span class="co">#&gt; Residuals 78   30066.3  385.47         0.99419       </span>
<span class="co">#&gt; Total     79   30242.1                 1.00000</span>

<span class="co">#4/ Run PERMDISP test (group dispersion in PCA1 x PCA2 space after excluding ancient samples) and assign results to object permdispR</span>
<span class="va">permdispR</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/smart_permdisp.html">smart_permdisp</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToGenoFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, sample_remove <span class="op">=</span> <span class="va">my_ancient</span><span class="op">)</span>
<span class="co">#assign sample summary to object permD</span>
<span class="va">permD</span> <span class="op">&lt;-</span><span class="va">permdispR</span><span class="op">$</span><span class="va">permdisp.samples</span>
<span class="co">#show PERMDISP table</span>
<span class="va">permdispR</span><span class="op">$</span><span class="va">permdisp.global_test</span>
<span class="co">#&gt;           Df      Sum Sq    Mean Sq         F Pr(&gt;F)</span>
<span class="co">#&gt; Groups     1  0.07254468 0.07254468 0.1911168 0.6693</span>
<span class="co">#&gt; Residuals 78 29.60747071 0.37958296        NA     NA</span>

<span class="co">#5/ Run PCA, PERMANOVA and PERMDISP in one run and assign results to object mvaR</span>
<span class="va">mvaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/smart_mva.html">smart_mva</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToGenoFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, sample_remove <span class="op">=</span> <span class="va">my_ancient</span><span class="op">)</span>
<span class="co"># assign statistical results to objects mvaR_eigen, mvaR_load and mvaR_coord</span>
<span class="va">mvaR_eigen</span> <span class="op">&lt;-</span> <span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.eigenvalues</span> <span class="co"># extract PCA eigenvalues</span>
<span class="va">mvaR_load</span> <span class="op">&lt;-</span> <span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.snp_loadings</span> <span class="co"># extract principal coefficients (SNP loadings)</span>
<span class="va">mvaR_coord</span> <span class="op">&lt;-</span> <span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.sample_coordinates</span> <span class="co"># extract PCA principal components (sample position in PCA space)</span>
<span class="co">#show PERMANOVA table</span>
<span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permanova.global_test</span>
<span class="co">#&gt;           Df SumsOfSqs MeanSqs F.Model      R2 Pr(&gt;F)</span>
<span class="co">#&gt; group      1     11849   11849 0.97217 0.01231 0.9092</span>
<span class="co">#&gt; Residuals 78    950644   12188         0.98769       </span>
<span class="co">#&gt; Total     79    962493                 1.00000</span>
<span class="co">#show PERMDISP table</span>
<span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permdisp.global_test</span> <span class="co"># extract PERMDISP table</span>
<span class="co">#&gt;           Df      Sum Sq    Mean Sq         F Pr(&gt;F)</span>
<span class="co">#&gt; Groups     1  0.07254468 0.07254468 0.1911168 0.6661</span>
<span class="co">#&gt; Residuals 78 29.60747071 0.37958296        NA     NA</span>
<span class="co">#assign sample summary to object mvaS</span>
<span class="va">mvaS</span> <span class="op">&lt;-</span> <span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test_samples</span>

<span class="co">#NOTE 1: Modify argument pc_axes to set the number of computed PCA axes (defaults: pc_axes = 2, program_svd = "RSpectra")</span>
<span class="co">#use program_svd = "bootSVD" for computing all PCA axes, where pc_axes has no effect on computations</span>
<span class="co">#NOTE 2: Missing values in dataset can only be coded as 9 (default: missing_value = 9) or NA (missing_value = NA)</span>
<span class="co">#SNPs with missing values are removed by default (missing_impute = "remove")</span>
<span class="co">#use missing_impute = "mean" for imputing missing values with SNP means </span>
<span class="co">#NOTE 3: arguments sample_remove and snp_remove remove any set of samples (by column number) and SNPs (by row number), respectively</span>
<span class="co">#defaults: sample_remove = FALSE, snp_remove = FALSE</span>
<span class="co">#NOTE 4: use argument sample_project to specify ancient samples by row number (default: sample_project = FALSE)</span>
<span class="co">#ancient samples are assumed to include missing values</span>
<span class="co">#if specified, ancient samples are always removed from PCA, PERMANOVA and PERMDISP computations</span>
<span class="co">#use argument pc_project to set the PCA space onto which ancient samples are projected (default: pc_project = c(1:2) for PCA 1 x PCA2 space)</span>

<span class="co">#6/ Plot PCA 1 x PCA 2</span>
<span class="co">#create colors for samples groups</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span>
<span class="co">#create color vector (group A = red, group B = blue, ancient samples = black)</span>
<span class="va">my_groups</span><span class="op">[</span><span class="va">my_ancient</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ancient"</span>; <span class="va">cols</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span>
<span class="co">#plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pcaR</span><span class="op">$</span><span class="va">pca.sample_coordinates</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>,<span class="st">"PC2"</span><span class="op">)</span><span class="op">]</span>, cex <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">19</span>, main <span class="op">=</span> <span class="st">"genotype smartpca"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">)</span>,  cex <span class="op">=</span> <span class="fl">1</span>, pch <span class="op">=</span> <span class="fl">19</span>, col <span class="op">=</span> <span class="va">cols</span>, text.col <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-example-1.png" width="100%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=smartsnp">https://​cloud.r-project.org/​package=smartsnp</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Salvador Herrando-Perez <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-6052-6854" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Christian Huber <br><small class="roles"> Contributor, maintainer </small> <a href="https://orcid.org/0000-0002-2267-2604" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
