<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Smart Principal Component Analysis — smart_pca • smartsnp</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Smart Principal Component Analysis — smart_pca" />
<meta property="og:description" content="Compute Principal Component Analysis (PCA) for variable x sample genotype data including covariance (centered), correlation (z-score) and SMARTPCA scaling,
and implements projection of ancient samples onto modern PCA space. SMARTPCA scaling controls for genetic drift when variables are bi-allelic genetic markers
such as single nucleotide polymorphisms (SNP) following Patterson, Price and Reich (2006).
Optimized to run fast single value decomposition for big datasets." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="../articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="../articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Smart Principal Component Analysis</h1>
    
    <div class="hidden name"><code>smart_pca.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compute Principal Component Analysis (PCA) for variable x sample genotype data including covariance (<code>centered</code>), correlation (z-score) and SMARTPCA scaling,
and implements projection of ancient samples onto modern PCA space. SMARTPCA scaling controls for genetic drift when variables are bi-allelic genetic markers
such as single nucleotide polymorphisms (SNP) following Patterson, Price and Reich (2006).
Optimized to run fast single value decomposition for big datasets.</p>
    </div>


    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>snp_data</th>
      <td><p>File name read from working directory.
SNP = rows, samples = columns without row names or column headings.
SNP values must be count data (no decimals allowed). File extension detected automatically whether text or <code>EIGENSTRAT</code>.
See details.</p></td>
    </tr>
    <tr>
      <th>packed_data</th>
      <td><p>Logical value for <code>EIGENSTRAT</code>, irrelevant for text data.
Default <code>packed_data = FALSE</code> assumes uncompressed <code>EIGENSTRAT</code>.
<code>packed_data = TRUE</code> for compressed or binary <code>EIGENSTRAT</code> (<code>PACKENDANCESTRYMAP</code>).</p></td>
    </tr>
    <tr>
      <th>sample_group</th>
      <td><p>Character or numeric vector assigning samples to groups.
Coerced to factor.</p></td>
    </tr>
    <tr>
      <th>sample_remove</th>
      <td><p>Logical <code>FALSE</code> or numeric vector indicating column numbers (samples) to be removed from computations.
Default <code>sample_remove =  FALSE</code> keeps all samples.</p></td>
    </tr>
    <tr>
      <th>snp_remove</th>
      <td><p>Logical <code>FALSE</code> or numeric vector indicating row numbers (SNPs) to be removed from computations.
Default <code>snp_remove =  FALSE</code> keeps all SNPs.
See details.</p></td>
    </tr>
    <tr>
      <th>missing_value</th>
      <td><p>Number <code>9</code> or string <code>NA</code> indicating missing value.
Default <code>missing_value = 9</code> as in <code>EIGENSTRAT</code>.
If no missing values present, no effect on computation.</p></td>
    </tr>
    <tr>
      <th>missing_impute</th>
      <td><p>String handling missing values.
Default <code>missing_impute = "mean"</code> replaces missing values of each SNP by mean of non-missing values across samples.
<code>missing_impute = "remove"</code> removes SNPs with at least one missing value.
If no missing values present, no effect on computation.</p></td>
    </tr>
    <tr>
      <th>scaling</th>
      <td><p>String. Default <code>scaling = "drift"</code> scales SNPs to control for expected allele frequency dispersion caused by genetic drift (SMARTPCA).
<code>scaling = "center"</code> for <code>centering</code> (covariance-based PCA).
<code>scaling = "sd"</code> for <code>centered</code> SNPs divided by standard deviation (correlation-based PCA).
<code>scaling = "none"</code> for no scaling.
See details.</p></td>
    </tr>
    <tr>
      <th>program_svd</th>
      <td><p>String indicating R package computing single value decomposition (SVD).
Default <code>program_svd = "Rspectra"</code> for <code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code>.
<code>program_svd = "bootSVD"</code> for <code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code>.
See details.</p></td>
    </tr>
    <tr>
      <th>pc_axes</th>
      <td><p>A numeric value.
If <code>program_svd = "Rspectra"</code> this argument indicates number of PCA axes computed starting with PCA axis 1.
Default <code>pc_axes = 2</code> computes PCA axes 1 and 2.
No effect on computation if <code>program_svd = "bootSVD"</code> since all PCA axes are computed.</p></td>
    </tr>
    <tr>
      <th>sample_project</th>
      <td><p>Numeric vector indicating column numbers (ancient samples) projected onto (modern) PCA space.
Default <code>sample_project =  FALSE</code> indicates no samples will be used for projection.
See details.</p></td>
    </tr>
    <tr>
      <th>pc_project</th>
      <td><p>Numeric vector indicating the ranks of the PCA axes ancient samples are projected onto.
Default <code>pc_ancient = c(1, 2)</code> for PCA axes 1 and 2. If <code>program_svd = "RSpectra"</code>, <code><a href='https://rdrr.io/r/base/length.html'>length(pc_ancient)</a></code> must be smaller than or equal to <code>pc_axes</code>.
No effect on computation, if no ancient samples present.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a list containing the following elements:</p><ul>
<li><p><code>pca.snp_loadings</code> Dataframe of principal coefficients of SNPs. One set of coefficients per PCA axis computed.<br /></p></li>
<li><p><code>pca.eigenvalues</code> Dataframe of eigenvalues, variance and cumulative variance explained. One eigenvalue per PCA axis computed.<br /></p></li>
<li><p><code>pca_sample_coordinates</code> Dataframe showing PCA sample summary.
  Column <em>Group</em> assigns samples to groups.
  Column <em>Class</em> specifies if samples "Removed" from PCA or "Projected" onto PCA space.
  Sequence of additional columns shows principal components (coordinates) of samples in PCA space (1 column per PCA computed named PC1, PC2, ...).</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>PCA is a rigid rotation of a Cartesian coordinate system (samples = points, axes = variables or SNPs) that maximizes the dispersion of points along a new system of axes (Pearson 1901; Hotelling 1933; Jolliffe 2002).
In rotated space (ordination), axes are <code>principal axes</code> (PCA axes), <code>eigenvalues</code> measure variance explained, and <code>principal coefficients</code> measure importance of SNPs (eigenvectors), <code>principal components</code> are coordinates of samples (i.e., linear combinations of scaled variables weighted by eigenvectors).
Principal coefficients are direction cosines between original and PCA axes (Legendre &amp; Legendre 2012). PCA can be computed by <code>eigenanalysis</code> or, as implemented here, single value decomposition (SVD). <br /></p>
<p>SNPs can be scaled in four different ways prior to SVD: (1) no scaling; (2) covariance: SNPs <code>centered</code> such that <em>M(i,j)</em> = <em>C(i,j)</em> minus <em>mean(j)</em>) where <em>C(i,j)</em> is the number of variant alleles for SNP <em>j</em> and sample <em>i</em>, and <em>M(i,j)</em> is the <code>centered</code> value of each data point; (3) correlation (z-scores): SNPs <code>centered</code> then divided by standard deviation <em>sd(j)</em>, (4) SMARTPCA: SNPs <code>centered</code> then divided by <em>sqrt(p(j)(1-p(j)))</em>, where <em>p(j)</em> equals <em>mean(j)</em> divided by <em>2</em>, quantifies the underlying allele frequency (autosomal chromosomes) and conceptualizes that SNP frequency changes at rate proportional to <em>sqrt(p(j)(1-p(j)))</em> per generation due to genetic drift (Patterson, Price and Reich 2006).
SMARTPCA standardization results in all SNPs that comply with Hardy-Weinberg equilibrium having identical variance.
SMARTPCA (Patterson, Price and Reich 2006) and <code>EIGENSTRAT</code> (Price, Patterson, Plenge, Weinblatt, Shadick and Reich 2006) are the computing suites of software <code>EIGENSOFT</code> (<a href='https://reich.hms.harvard.edu/software'>https://reich.hms.harvard.edu/software</a>).<br /></p>
<p><code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code> runs single value decomposition much faster than <code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code>. With <code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code>, <code>pc_axes</code> indicates number of eigenvalues and eigenvectors computed starting from PCA axis 1. <code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code> computes all eigenvalues and eigenvectors. Eigenvalues calculated from singular values divided by number of samples minus 1. If number of samples equals number of SNPS, <code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code> prints message alert that no computing efficiency is achieved for square matrices.<br /></p>
<p>Ancient samples (with many missing values) can be projected onto modern PCA space derived from modern samples.
Following Nelson Taylor and MacGregor (1996), the projected coordinates of a given ancient sample equal the slope coefficient of linear fit through the origin of (scaled) non-missing SNP values of that sample (response) versus principal coefficients of same SNPs in modern samples.
Number of projected coordinates per ancient sample given by <code><a href='https://rdrr.io/r/base/length.html'>length(pc_ancient)</a></code>.
With <code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code>, <code>pc_axes</code> must be larger or equal to <code><a href='https://rdrr.io/r/base/length.html'>length(pc_ancient)</a></code>.<br /></p>
<p>Data read from working directory with SNPs as rows and samples as columns.
Two alternative formats: (1) text file of SNPs by samples (file extension and column separators recognized automatically) read using <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code>; or (2) duet of <code>EIGENSTRAT</code> files (see <a href='https://reich.hms.harvard.edu/software'>https://reich.hms.harvard.edu/software</a>) using <code><a href='https://vroom.r-lib.org/reference/vroom_fwf.html'>vroom_fwf</a></code>, including a genotype file of SNPs by samples (<code>*.geno</code>), and a sample file (<code>*.ind</code>) containing three vectors assigning individual samples to unique user-predefined groups (populations), sexes (or other user-defined descriptor) and alphanumeric identifiers.
For <code>EIGENSTRAT</code>, vector <code>sample_group</code> assigns samples to groups retrievable from column of file <code>*.ind</code>. SNPs with zero variance removed prior to SVD to optimize computation time and avoid undefined values if <code>scaling = "sd"</code> or <code>"drift"</code>.<br /></p>
<p>Users can select subsets of samples or SNPs by introducing a vector including column numbers for samples (<code>sample_remove</code>) and/or row numbers for SNPs (<code>snp_remove</code>) to be removed from computations.
Function stops if the final number of SNPs is 1 or 2.
<code>EIGENSOFT</code> was conceived for the analysis of human genes and its SMARTPCA suite so accepts 22 (autosomal) chromosomes by default.
If &gt;22 chromosomes are provided and the internal parameter <code>numchrom</code> is not set to the target number chromosomes of interest, SMARTPCA automatically subsets chromosomes 1 to 22.
In contrast, <code>smart_pca</code> accepts any number of autosomes with or without the sex chromosomes from an <code>EIGENSTRAT</code> file.<br /></p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Hotelling, H. (1933) Analysis of a complex of statistical variables into principal components. Journal of Educational Psychology, 24, 417-441.<br /></p>
<p>Jolliffe, I.T. (2002) Principal Component Analysis (Springer, New York, USA).<br /></p>
<p>Legendre, P. &amp; L. F. J. Legendre (2012). Numerical ecology.  Developments in environmental modelling (Elsevier, Oxford, UK).<br /></p>
<p>Nelson, P.R.C., P.A. Taylor, and J.F. MacGregor (1996) Missing data methods in PCA and PLS: score calculations with incomplete observations. Chemometrics and Intelligent Laboratory Systems, 35, 45-65.<br /></p>
<p>Patterson, N.J., A. L. Price and D. Reich (2006) Population structure and eigenanalysis. PLoS Genetics, 2, e190.<br /></p>
<p>Pearson, K. (1901) On lines and planes of closest fit to systems of points in space. Philosophical Magazine, 2, 559-572.<br /></p>
<p>Price, A.L., N.J. Patterson, R.M. Plenge, M.E. Weinblatt, N.A. Shadick and David Reich (2006). Principal components analysis corrects for stratification in genome-wide association studies. Nature Genetics, 38, 904-909.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code> (package <b>bootSVD</b>),
<code><a href='https://rdrr.io/pkg/foreach/man/foreach.html'>foreach</a></code> (package <b>foreach</b>),
<code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code> (package <b>data.table</b>),
<code><a href='https://rdrr.io/pkg/Rfast/man/colVars.html'>rowVars</a></code> (package <b>Rfast</b>),
<code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code> (package <b>RSpectra</b>),
<code><a href='https://vroom.r-lib.org/reference/vroom_fwf.html'>vroom_fwf</a></code> (package <b>vroom</b>)</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Path to example genotype matrix "dataSNP"</span>
<span class='va'>pathToGenoFile</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"dataSNP"</span>, package <span class='op'>=</span> <span class='st'>"smartsnp"</span><span class='op'>)</span>

<span class='co'># Example 1: modern samples</span>
<span class='co'>#assign 50 samples to each of two groups and colors</span>
<span class='va'>my_groups</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"A"</span>, <span class='fl'>50</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"B"</span>, <span class='fl'>50</span><span class='op'>)</span><span class='op'>)</span>; <span class='va'>cols</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"red"</span>, <span class='st'>"blue"</span><span class='op'>)</span>
<span class='co'>#run PCA with truncated SVD (PCA 1 x PCA 2)</span>
<span class='va'>pcaR1</span> <span class='op'>&lt;-</span> <span class='fu'>smart_pca</span><span class='op'>(</span>snp_data <span class='op'>=</span> <span class='va'>pathToGenoFile</span>, sample_group <span class='op'>=</span> <span class='va'>my_groups</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Checking argument options selected...</span></div><div class='output co'>#&gt; <span class='message'>Argument options are correct...</span></div><div class='output co'>#&gt; <span class='message'>Loading data...</span></div><div class='output co'>#&gt; <span class='message'>Imported 10000 SNP by 100 sample genotype matrix</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Filtering data...</span></div><div class='output co'>#&gt; <span class='message'>No samples projected after PCA computation</span></div><div class='output co'>#&gt; <span class='message'>10000 SNPs included in PCA computation</span></div><div class='output co'>#&gt; <span class='message'>1 SNPs omitted from PCA computation</span></div><div class='output co'>#&gt; <span class='message'>100 samples included in PCA computation</span></div><div class='output co'>#&gt; <span class='message'>Completed data filtering</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scanning for invariant SNPs...</span></div><div class='output co'>#&gt; <span class='message'>Scan complete: no invariant SNPs found</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Checking for missing values...</span></div><div class='output co'>#&gt; <span class='message'>6341 SNPs contain missing values</span></div><div class='output co'>#&gt; <span class='message'>Imputing SNPs with missing values...</span></div><div class='output co'>#&gt; <span class='message'>Imputation with means completed: 10058 missing values imputed</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scaling values by SNP...</span></div><div class='output co'>#&gt; <span class='message'>Centering and scaling by drift dispersion...</span></div><div class='output co'>#&gt; <span class='message'>Completed scaling using drift</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Computing singular value decomposition using RSpectra...</span></div><div class='output co'>#&gt; <span class='message'>Completed singular value decomposition using RSpectra</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Extracting eigenvalues and eigenvectors...</span></div><div class='output co'>#&gt; <span class='message'>Eigenvalues and eigenvectors extracted</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Tabulating PCA outputs...</span></div><div class='output co'>#&gt; <span class='message'>Completed tabulations of PCA outputs...</span></div><div class='input'><span class='va'>pcaR1</span><span class='op'>$</span><span class='va'>pca.eigenvalues</span> <span class='co'># extract eigenvalues</span>
</div><div class='output co'>#&gt;                                      PC1        PC2
#&gt; observed eigenvalues          321.195060 317.773982
#&gt; variance explained              1.209482   1.196600
#&gt; cumulative variance explained   1.209482   2.406083</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>pcaR1</span><span class='op'>$</span><span class='va'>pca.snp_loadings</span><span class='op'>)</span> <span class='co'># extract principal coefficients (SNP loadings)</span>
</div><div class='output co'>#&gt;             PC1          PC2
#&gt; 1  0.0046337236 -0.002775738
#&gt; 2  0.0018116572  0.011950491
#&gt; 3  0.0096288656 -0.011907379
#&gt; 4  0.0055371026  0.001682321
#&gt; 5 -0.0046772139  0.002187803
#&gt; 6  0.0004521752  0.002215284</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>pcaR1</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>)</span> <span class='co'># extract principal components (sample position in PCA space)</span>
</div><div class='output co'>#&gt;   Group Class        PC1        PC2
#&gt; 1     A   PCA   4.768216  13.413303
#&gt; 2     A   PCA -16.661893  -9.789570
#&gt; 3     A   PCA -25.643492   5.994816
#&gt; 4     A   PCA  -2.453762 -25.043319
#&gt; 5     A   PCA  19.760752   7.214635
#&gt; 6     A   PCA -28.587658   2.656167</div><div class='input'><span class='co'>#plot PCA</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>pcaR1</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>[</span>,<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PC1"</span>,<span class='st'>"PC2"</span><span class='op'>)</span><span class='op'>]</span>, cex <span class='op'>=</span> <span class='fl'>2</span>,
     pch <span class='op'>=</span> <span class='fl'>19</span>, col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>as.factor</a></span><span class='op'>(</span><span class='va'>my_groups</span><span class='op'>)</span><span class='op'>]</span>, main <span class='op'>=</span> <span class='st'>"genotype smartpca"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topleft"</span>, legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>as.factor</a></span><span class='op'>(</span><span class='va'>my_groups</span><span class='op'>)</span><span class='op'>)</span>, cex <span class='op'>=</span><span class='fl'>1</span>,
     pch <span class='op'>=</span> <span class='fl'>19</span>, col <span class='op'>=</span> <span class='va'>cols</span>, text.col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>)</span>
</div><div class='img'><img src='smart_pca-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Example 2: modern and ancient samples (ancient samples projected onto modern PCA space)</span>
<span class='co'>#assign samples 1st to 10th per group to ancient</span>
<span class='va'>my_ancient</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>, <span class='fl'>51</span><span class='op'>:</span><span class='fl'>60</span><span class='op'>)</span>
<span class='co'>#run PCA with truncated SVD (PCA 1 x PCA 2)</span>
<span class='va'>pcaR2</span> <span class='op'>&lt;-</span> <span class='fu'>smart_pca</span><span class='op'>(</span>snp_data <span class='op'>=</span> <span class='va'>pathToGenoFile</span>, sample_group <span class='op'>=</span> <span class='va'>my_groups</span>, sample_project <span class='op'>=</span> <span class='va'>my_ancient</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Checking argument options selected...</span></div><div class='output co'>#&gt; <span class='message'>Argument options are correct...</span></div><div class='output co'>#&gt; <span class='message'>Loading data...</span></div><div class='output co'>#&gt; <span class='message'>Imported 10000 SNP by 100 sample genotype matrix</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Filtering data...</span></div><div class='output co'>#&gt; <span class='message'>10000 SNPs included in PCA computation</span></div><div class='output co'>#&gt; <span class='message'>1 SNPs omitted from PCA computation</span></div><div class='output co'>#&gt; <span class='message'>80 samples included in PCA computation</span></div><div class='output co'>#&gt; <span class='message'>20 samples projected after PCA computation</span></div><div class='output co'>#&gt; <span class='message'>Completed data filtering</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scanning for invariant SNPs...</span></div><div class='output co'>#&gt; <span class='message'>Scan complete: no invariant SNPs found</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Checking for missing values...</span></div><div class='output co'>#&gt; <span class='message'>5468 SNPs contain missing values</span></div><div class='output co'>#&gt; <span class='message'>Imputing SNPs with missing values...</span></div><div class='output co'>#&gt; <span class='message'>Imputation with means completed: 7946 missing values imputed</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scaling values by SNP...</span></div><div class='output co'>#&gt; <span class='message'>Centering and scaling by drift dispersion...</span></div><div class='output co'>#&gt; <span class='message'>Completed scaling using drift</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Computing singular value decomposition using RSpectra...</span></div><div class='output co'>#&gt; <span class='message'>Completed singular value decomposition using RSpectra</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Extracting eigenvalues and eigenvectors...</span></div><div class='output co'>#&gt; <span class='message'>Eigenvalues and eigenvectors extracted</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Projecting ancient samples onto PCA space</span></div><div class='output co'>#&gt; <span class='message'>PCA space = PC1PC2</span></div><div class='output co'>#&gt; <span class='message'>Completed ancient sample projection</span></div><div class='output co'>#&gt; <span class='message'>20 ancient samples projected</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Tabulating PCA outputs...</span></div><div class='output co'>#&gt; <span class='message'>Completed tabulations of PCA outputs...</span></div><div class='input'><span class='va'>pcaR2</span><span class='op'>$</span><span class='va'>pca.eigenvalues</span> <span class='co'># extract eigenvalues</span>
</div><div class='output co'>#&gt;                                      PC1        PC2
#&gt; observed eigenvalues          395.522794 390.860709
#&gt; variance explained              1.486341   1.468821
#&gt; cumulative variance explained   1.486341   2.955162</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>pcaR2</span><span class='op'>$</span><span class='va'>pca.snp_loadings</span><span class='op'>)</span> <span class='co'># extract principal coefficients (SNP loading)</span>
</div><div class='output co'>#&gt;             PC1           PC2
#&gt; 1 -0.0096014176 -0.0009125866
#&gt; 2  0.0056122566  0.0051110423
#&gt; 3 -0.0009022379 -0.0120422363
#&gt; 4  0.0111802251 -0.0086896301
#&gt; 5 -0.0023697339  0.0102539820
#&gt; 6  0.0041931801  0.0008440232</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>pcaR2</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>)</span> <span class='co'># extract principal components (sample position in PCA space)</span>
</div><div class='output co'>#&gt;   Group     Class        PC1        PC2
#&gt; 1     A Projected  0.7337079 -0.4975136
#&gt; 2     A Projected -1.6726992 -1.1502356
#&gt; 3     A Projected -0.5604441 -2.0200176
#&gt; 4     A Projected -3.0152112 -2.1284274
#&gt; 5     A Projected  3.1316463 -2.6935686
#&gt; 6     A Projected -1.9680503 -0.5368760</div><div class='input'><span class='co'>#assign samples to groups (A, ancient, B) and colors</span>
<span class='va'>my_groups</span><span class='op'>[</span><span class='va'>my_ancient</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"ancient"</span>; <span class='va'>cols</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"red"</span>, <span class='st'>"black"</span>, <span class='st'>"blue"</span><span class='op'>)</span>
<span class='co'>#plot PCA</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>pcaR2</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>[</span>,<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PC1"</span>,<span class='st'>"PC2"</span><span class='op'>)</span><span class='op'>]</span>,
     cex <span class='op'>=</span> <span class='fl'>2</span>, col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>as.factor</a></span><span class='op'>(</span><span class='va'>my_groups</span><span class='op'>)</span><span class='op'>]</span>, pch <span class='op'>=</span> <span class='fl'>19</span>, main <span class='op'>=</span> <span class='st'>"genotype smartpca"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topleft"</span>, legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>as.factor</a></span><span class='op'>(</span><span class='va'>my_groups</span><span class='op'>)</span><span class='op'>)</span>,  cex <span class='op'>=</span> <span class='fl'>1</span>,
     pch <span class='op'>=</span> <span class='fl'>19</span>, col <span class='op'>=</span> <span class='va'>cols</span>, text.col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>)</span>
</div><div class='img'><img src='smart_pca-2.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


