<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Projecting ancient samples • smartsnp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Projecting ancient samples">
<meta property="og:description" content="This Vignette provides an example of how to project ancient DNA onto modern data using  the smartsnp package.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="../articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="../articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ChristianHuber/smartsnp/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="aDNA_smartpca_analysis_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Projecting ancient samples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ChristianHuber/smartsnp/blob/master/vignettes/aDNA_smartpca_analysis.Rmd"><code>vignettes/aDNA_smartpca_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>aDNA_smartpca_analysis.Rmd</code></div>

    </div>

    
    
<p>This Vignette provides an example of projecting ancient DNA onto modern data in a PCA analysis using the <em>smartsnp</em> package. We will use data from one of the first large-scale ancient DNA studies, Lazaridis et al. 2016:</p>
<p>Lazaridis et al. “Genomic insights into the origin of farming in the ancient Near East”, Nature volume 536, pages 419–424 (2016).</p>
<p>The data is available online but needs to be pre-processed. Particularly, the aDNA data needs to be merged with modern data. All steps can be completed in R, but some need certain command-line software installed - they won’t work on a Windows machine.</p>
<p>If you are just interested in how to run <em>smart_pca</em> with ancient samples (all you need is an index vector <em>aDNA_inds</em> with the column numbers of the ancient samples), feel free to go straight to the section “Running smartpca” below.</p>
<div id="install-package-smartsnp" class="section level2">
<h2 class="hasAnchor">
<a href="#install-package-smartsnp" class="anchor"></a>Install package <em>smartsnp</em>
</h2>
<p>Select one of two options.</p>
<p>Install development version from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"ChristianHuber/smartsnp"</span><span class="op">)</span></code></pre></div>
<p>Install release version from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"smartsnp"</span><span class="op">)</span></code></pre></div>
<p>Load the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://christianhuber.github.io/smartsnp/">smartsnp</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="downloading-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-the-data" class="anchor"></a>Downloading the data</h2>
<p>First, set the working directory to a location where you want to download and process the files. In my case, I’m choosing the Downloads directory in my home folder.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"~/Downloads/"</span><span class="op">)</span></code></pre></div>
<p>We will download the data provided here <a href="https://reich.hms.harvard.edu/datasets" class="uri">https://reich.hms.harvard.edu/datasets</a> using a command-line software called <em>wget</em>. Alternatively, you can download the file using a browser and the link. Note that this is quite a large file, &gt;200 Mb!</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"wget https://reich.hms.harvard.edu/sites/reich.hms.harvard.edu/files/inline-files/NearEastPublic.tar.gz"</span><span class="op">)</span></code></pre></div>
<p>The downloaded data has to be unzipped. I will unzip it into a new folder called “data”.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"mkdir data"</span><span class="op">)</span>                               <span class="co"># Make a new folder called "data"</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"tar -xvf NearEastPublic.tar.gz -C ./data"</span><span class="op">)</span> <span class="co"># Unzip data into this folder</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"rm NearEastPublic.tar.gz"</span><span class="op">)</span>                 <span class="co"># Remove the zip file</span></code></pre></div>
</div>
<div id="select-subsets-of-individuals-with-convertf" class="section level2">
<h2 class="hasAnchor">
<a href="#select-subsets-of-individuals-with-convertf" class="anchor"></a>Select subsets of individuals with convertf</h2>
<p>The data is in a PACKEDANCESTRYMAP format. The ancient and modern data is in two different files, and we are only interested in the Western Eurasian subset of the modern samples.</p>
<p>In the next step, we will filter out the Western Eurasian samples from the full set of modern samples. Then, we will merge the modern with the ancient data.</p>
<p>The <em>convertf</em> and <em>mergeit</em> command-line software of the Eigensoft package conveniently allows to run these two operations, see here for installation: <a href="https://github.com/DReichLab/EIG" class="uri">https://github.com/DReichLab/EIG</a></p>
<p>To run Eigensoft software within R using the <em>system</em> function, you might need to first explicitly tell R where the software can be found:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Path to Eigensoft binaries (might be different on your computer):</span>
<span class="va">pathToEIGENSOFT</span> <span class="op">=</span> <span class="st">"~/repos/EIG/bin/"</span>
<span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span><span class="op">(</span>PATH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html">Sys.getenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="st">"PATH"</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">":"</span>, <span class="va">pathToEIGENSOFT</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We need to generate a parameter file for <em>convertf</em> that contains all the file names. We also need to generate a list of West Eurasian populations in a text file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Generating a text file with West Eurasian group names</span>
<span class="va">westEurasian_pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"Abkhasian"</span>, <span class="st">"Adygei"</span>, <span class="st">"Albanian"</span>, <span class="st">"Armenian"</span>, <span class="st">"Assyrian"</span>, <span class="st">"Balkar"</span>, <span class="st">"Basque"</span>, <span class="st">"BedouinA"</span>, <span class="st">"BedouinB"</span>, <span class="st">"Belarusian"</span>, <span class="st">"Bulgarian"</span>, <span class="st">"Canary_Islander"</span>, 
  <span class="st">"Chechen"</span>, <span class="st">"Croatian"</span>, <span class="st">"Cypriot"</span>, <span class="st">"Czech"</span>, <span class="st">"Druze"</span>, <span class="st">"English"</span>, <span class="st">"Estonian"</span>, <span class="st">"Finnish"</span>, <span class="st">"French"</span>, <span class="st">"Georgian"</span>, <span class="st">"German"</span>, <span class="st">"Greek"</span>, <span class="st">"Hungarian"</span>, <span class="st">"Icelandic"</span>, 
  <span class="st">"Iranian"</span>, <span class="st">"Irish"</span>, <span class="st">"Irish_Ulster"</span>, <span class="st">"Italian_North"</span>, <span class="st">"Italian_South"</span>, <span class="st">"Jew_Ashkenazi"</span>, <span class="st">"Jew_Georgian"</span>, <span class="st">"Jew_Iranian"</span>, <span class="st">"Jew_Iraqi"</span>, <span class="st">"Jew_Libyan"</span>, <span class="st">"Jew_Moroccan"</span>, 
  <span class="st">"Jew_Tunisian"</span>, <span class="st">"Jew_Turkish"</span>, <span class="st">"Jew_Yemenite"</span>, <span class="st">"Jordanian"</span>, <span class="st">"Kumyk"</span>, <span class="st">"Lebanese_Christian"</span>, <span class="st">"Lebanese"</span>, <span class="st">"Lebanese_Muslim"</span>, <span class="st">"Lezgin"</span>, <span class="st">"Lithuanian"</span>, <span class="st">"Maltese"</span>, 
  <span class="st">"Mordovian"</span>, <span class="st">"North_Ossetian"</span>, <span class="st">"Norwegian"</span>, <span class="st">"Orcadian"</span>, <span class="st">"Palestinian"</span>, <span class="st">"Polish"</span>, <span class="st">"Romanian"</span>, <span class="st">"Russian"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Saudi"</span>, <span class="st">"Scottish"</span>, <span class="st">"Shetlandic"</span>, <span class="st">"Sicilian"</span>, 
  <span class="st">"Sorb"</span>, <span class="st">"Spanish_North"</span>, <span class="st">"Spanish"</span>, <span class="st">"Syrian"</span>, <span class="st">"Turkish"</span>, <span class="st">"Ukrainian"</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Generating the parameter file for convertf:</span>
<span class="va">par.ANCESTRYMAP.FILTER</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"genotypename:    ./data/HumanOriginsPublic2068.geno"</span>,
  <span class="st">"snpname:         ./data/HumanOriginsPublic2068.snp"</span>,
  <span class="st">"indivname:       ./data/HumanOriginsPublic2068.ind"</span>,
  <span class="st">"poplistname: ./WestEurasia.poplist.txt"</span>,
  <span class="st">"genotypeoutname: ./data/HumanOriginsPublic2068.WestEurasia.geno"</span>,
  <span class="st">"snpoutname:      ./data/HumanOriginsPublic2068.WestEurasia.snp"</span>,
  <span class="st">"indivoutname:    ./data/HumanOriginsPublic2068.WestEurasia.ind"</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/writeLines.html">writeLines</a></span><span class="op">(</span><span class="va">par.ANCESTRYMAP.FILTER</span>, con <span class="op">=</span> <span class="st">"par.ANCESTRYMAP.FILTER"</span><span class="op">)</span>

<span class="co"># Now run convertf using the system command in R. This is equivalent to running the quoted command in a terminal:</span>

<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"convertf -p par.ANCESTRYMAP.FILTER"</span><span class="op">)</span></code></pre></div>
</div>
<div id="merging-ancient-with-modern-data-using-mergeit" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-ancient-with-modern-data-using-mergeit" class="anchor"></a>Merging ancient with modern data using mergeit</h2>
<p>Now we combine the ancient samples with the modern data using <em>mergeit</em>. Again, we first need a parameter file and then we can run <em>mergeit</em> with the <em>system</em> function in R (or alternatively in the terminal).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"geno1: ./data/HumanOriginsPublic2068.WestEurasia.geno"</span>,
  <span class="st">"snp1:  ./data/HumanOriginsPublic2068.WestEurasia.snp"</span>,
  <span class="st">"ind1:  ./data/HumanOriginsPublic2068.WestEurasia.ind"</span>,
  <span class="st">"geno2: ./data/AncientLazaridis2016.geno"</span>,
  <span class="st">"snp2:  ./data/AncientLazaridis2016.snp"</span>,
  <span class="st">"ind2:  ./data/AncientLazaridis2016.ind"</span>,
  <span class="st">"genooutfilename: ./data/AncientLazaridis2016_ModernWestEurasia.geno"</span>,
  <span class="st">"snpoutfilename:  ./data/AncientLazaridis2016_ModernWestEurasia.snp"</span>,
  <span class="st">"indoutfilename:  ./data/AncientLazaridis2016_ModernWestEurasia.ind"</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/writeLines.html">writeLines</a></span><span class="op">(</span><span class="va">params</span>, con <span class="op">=</span> <span class="st">"mergeit.params.txt"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"mergeit -p mergeit.params.txt"</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-smartpca" class="section level2">
<h2 class="hasAnchor">
<a href="#running-smartpca" class="anchor"></a>Running smartpca</h2>
<p>We still need two additional vectors before we can run <em>smartsnp</em>: one that defines the ancient samples, and one that defines which samples we want to remove before running the PCA.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Group names of the ancient groups</span>
<span class="va">aDNA_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Anatolia_ChL"</span>, <span class="st">"Anatolia_N"</span>, <span class="st">"Armenia_ChL"</span>, <span class="st">"Armenia_EBA"</span>, <span class="st">"Armenia_MLBA"</span>, <span class="st">"CHG"</span>, <span class="st">"EHG"</span>, <span class="st">"Europe_EN"</span>, <span class="st">"Europe_LNBA"</span>, <span class="st">"Europe_MNChL"</span>, <span class="st">"Iberia_BA"</span>, <span class="st">"Iran_ChL"</span>, <span class="st">"Iran_HotuIIIb"</span>, <span class="st">"Iran_LN"</span>, <span class="st">"Iran_N"</span>, <span class="st">"Levant_BA"</span>, <span class="st">"Levant_N"</span>, <span class="st">"Natufian"</span>, <span class="st">"SHG"</span>, <span class="st">"Steppe_EMBA"</span>, <span class="st">"Steppe_Eneolithic"</span>, <span class="st">"Steppe_IA"</span>, <span class="st">"Steppe_MLBA"</span>, <span class="st">"Switzerland_HG"</span>, <span class="st">"WHG"</span><span class="op">)</span>

<span class="co"># Contains group names of all groups in merged data</span>
<span class="va">GR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"./data/AncientLazaridis2016_ModernWestEurasia.ind"</span>, header<span class="op">=</span><span class="cn">F</span><span class="op">)</span>

<span class="co"># Vector defining ancient and modern groups</span>
<span class="va">SA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">GR</span><span class="op">$</span><span class="va">V3</span> <span class="op">%in%</span> <span class="va">westEurasian_pops</span>, <span class="st">"modern"</span>, <span class="st">"ancient"</span><span class="op">)</span>

<span class="co"># Samples to remove:</span>
<span class="va">sample.rem</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mota"</span>, <span class="st">"Denisovan"</span>, <span class="st">"Chimp"</span>, <span class="st">"Mbuti.DG"</span>, <span class="st">"Altai"</span>, 
               <span class="st">"Vi_merge"</span>, <span class="st">"Clovis"</span>, <span class="st">"Kennewick"</span>, <span class="st">"Chuvash"</span>, <span class="st">"Ust_Ishim"</span>, 
               <span class="st">"AG2"</span>, <span class="st">"MA1"</span>, <span class="st">"MezE"</span>, <span class="st">"hg19ref"</span>, <span class="st">"Kostenki14"</span><span class="op">)</span>

<span class="co"># Simple index vectors that determines which samples to remove and which to use for PCA ordination or projection:</span>
<span class="va">SR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">GR</span><span class="op">$</span><span class="va">V3</span> <span class="op">%in%</span> <span class="va">sample.rem</span><span class="op">)</span>  <span class="co"># Column numbers of samples to remove</span>
<span class="va">SP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">SA</span> <span class="op">==</span> <span class="st">"ancient"</span><span class="op">)</span>        <span class="co"># Column numbers of samples to project (i.e. aDNA)</span></code></pre></div>
<p>Now we are finally ready to run smart_pca:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Running smart_pca:</span>
<span class="va">sm.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_pca.html">smart_pca</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="st">"./data/AncientLazaridis2016_ModernWestEurasia.geno"</span>,
                    sample_group <span class="op">=</span> <span class="va">GR</span><span class="op">$</span><span class="va">V3</span>, missing_value <span class="op">=</span> <span class="fl">9</span>, missing_impute <span class="op">=</span> <span class="st">"mean"</span>,
                    scaling <span class="op">=</span> <span class="st">"drift"</span>, program_svd <span class="op">=</span> <span class="st">"RSpectra"</span>, pc_axes <span class="op">=</span> <span class="fl">2</span>, 
                    sample_remove <span class="op">=</span> <span class="va">SR</span>, sample_project <span class="op">=</span> <span class="va">SP</span>, pc_project <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>

<span class="co"># To see more information on the different parameter options: </span>
<span class="op">?</span><span class="va">smart_pca</span></code></pre></div>
</div>
<div id="plotting-the-results-using-ggplot2" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-results-using-ggplot2" class="anchor"></a>Plotting the results using ggplot2</h2>
<p>We can have a look at the result using <em>ggplot2</em> (and <em>ggrepel</em> for labeling the groups). The <em>data.table</em> package is used to simplify some data operations. For more info on these packages, see:</p>
<p><a href="https://ggplot2.tidyverse.org/" class="uri">https://ggplot2.tidyverse.org/</a></p>
<p><a href="https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html" class="uri">https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html</a></p>
<p><a href="https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html" class="uri">https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html</a></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This needs the R libraries ggplot2, ggrepel, and data.table.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span>

<span class="co"># Plotting with ggplot2 and labeling the groups with ggrepel:</span>

<span class="va">smart_mva.aDNA_WestEurasia.evec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span><span class="va">sm.pca</span><span class="op">$</span><span class="va">pca.sample_coordinates</span><span class="op">)</span>
<span class="va">smart_mva.aDNA_WestEurasia.evec</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1mean"</span>, <span class="st">"PC2mean"</span><span class="op">)</span> <span class="op">:=</span> <span class="fu">.</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">PC1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">PC2</span><span class="op">)</span><span class="op">)</span>, <span class="va">Group</span><span class="op">]</span>
<span class="va">smart_mva.aDNA_WestEurasia.evec</span><span class="op">[</span>, <span class="va">Name</span> <span class="op">:=</span> <span class="va">GR</span><span class="op">$</span><span class="va">V1</span><span class="op">]</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">smart_mva.aDNA_WestEurasia.evec</span><span class="op">[</span><span class="va">Class</span> <span class="op">==</span> <span class="st">"PCA"</span><span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="op">-</span><span class="va">PC1</span>, <span class="va">PC2</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"grey"</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">smart_mva.aDNA_WestEurasia.evec</span><span class="op">[</span><span class="va">Class</span> <span class="op">==</span> <span class="st">"Projected"</span> <span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="op">-</span><span class="va">PC1</span>, <span class="va">PC2</span>, fill<span class="op">=</span><span class="va">Group</span>, shape<span class="op">=</span><span class="va">Group</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">21</span><span class="op">:</span><span class="fl">25</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">smart_mva.aDNA_WestEurasia.evec</span><span class="op">[</span><span class="va">Class</span> <span class="op">==</span> <span class="st">"Projected"</span>,<span class="va">.SD</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">Group</span><span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="op">-</span><span class="va">PC1mean</span>, <span class="va">PC2mean</span>, label<span class="op">=</span><span class="va">Group</span>, col<span class="op">=</span><span class="va">Group</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.7</span>, segment.color<span class="op">=</span><span class="st">"NA"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></code></pre></div>
<p><img src="figure/lazaridis_plot-1.png" title="plot of chunk lazaridis_plot" alt="plot of chunk lazaridis_plot" style="display: block; margin: auto;"></p>
<p>Voila! Note that we have plotted the negative of PC1 (i.e. -PC1) here. The only reason for this is to make the plot have the same orientation as the original plot in Lazaridis et al. (2016), Fig. 1B. Importantly, changing the sign of any axis of an PCA does not change its interpretation, and different software can give different signs. See the excellent explanation here:</p>
<p><a href="https://stats.stackexchange.com/questions/88880/does-the-sign-of-scores-or-of-loadings-in-pca-or-fa-have-a-meaning-may-i-revers" class="uri">https://stats.stackexchange.com/questions/88880/does-the-sign-of-scores-or-of-loadings-in-pca-or-fa-have-a-meaning-may-i-revers</a></p>
<p>Finally, let’s move back to the old working directory:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
