<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Example PCA, PERMANOVA and PERMDISP analysis • smartsnp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Example PCA, PERMANOVA and PERMDISP analysis">
<meta property="og:description" content="This Vignette provides an example analysis of genetic data using  the smartsnp package.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="../articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="../articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="mallard_smartpca_analysis_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Example PCA, PERMANOVA and PERMDISP analysis</h1>
            
      
      
      <div class="hidden name"><code>mallard_smartpca_analysis.Rmd</code></div>

    </div>

    
    
<p>This Vignette provides an example analysis of genetic data using the smartsnp package.</p>
<div id="description-of-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#description-of-the-data" class="anchor"></a>Description of the data</h2>
<p>Multivariate analysis of mallard genotypes using the dataset published by Kraus et al. 2013.</p>
<p>Paper = <a href="https://onlinelibrary.wiley.com/doi/10.1111/mec.12098" class="uri">https://onlinelibrary.wiley.com/doi/10.1111/mec.12098</a></p>
<p>Dataset = <a href="https://datadryad.org/stash/dataset/doi:10.5061/dryad.1bq39" class="uri">https://datadryad.org/stash/dataset/doi:10.5061/dryad.1bq39</a></p>
<p>Population SEAP removed from dataset as its geographic background is unclear (Robert Kraus, pers. comm., 02/06/2021).</p>
<p>Populations GBAB, GBFE and GBNM (British Isles) removed from dataset as these individuals might have mixed with captive/feral mallards (Robert Kraus, pers. comm., 03/06/2021).</p>
<p>Three datasets are available. They are part of the <em>smartsnp</em> package, you don’t need to download or process the data from dryad:</p>
<ul>
<li>Genotype data (mallard_genotype_Kraus2012.txt) = 364 SNPs (rows) x 695 individuals (columns), individuals comprise 55 populations and 10 flyways. Genotypes are 0, 1, 2 or (for missing values) 9</li>
<li>Group names (mallard_samples_Kraus2013.txt) = 695 rows x 3 columns, column 1 = flyway names, column 2 = population name, column 3 = indvidual names comprise 55 populations and 10 flyways</li>
<li>SNP names (mallard_snps_Kraus2013.txt) = 695 rows x 1 column</li>
</ul>
<p>The study supports panmixia in cosmopolitan bird species (see Kraus et al. 2013):</p>
<p>“…Only Greenland is genetically differentiated from the remaining mallard population, and to a lesser extent, slight differentiation is observed between flyways in Europe and North America”.</p>
<p>“…There is a lack of clear population structure, suggesting that the world’s mallards, perhaps with minor exceptions, form a single large, mainly interbreeding population”.</p>
</div>
<div id="install-package-smartsnp-use-one-option" class="section level2">
<h2 class="hasAnchor">
<a href="#install-package-smartsnp-use-one-option" class="anchor"></a>Install package <em>smartsnp</em> (use one option)</h2>
<p>From GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"ChristianHuber/smartsnp"</span><span class="op">)</span></code></pre></div>
<p>From CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"smartsnp"</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-package" class="section level2">
<h2 class="hasAnchor">
<a href="#load-package" class="anchor"></a>Load package</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">smartsnp</span><span class="op">)</span></code></pre></div>
</div>
<div id="create-group-factor" class="section level2">
<h2 class="hasAnchor">
<a href="#create-group-factor" class="anchor"></a>Create group factor</h2>
<p>Load group file (flyway = categorical predictor in PERMANOVA AND PERMDISP tests):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_samples_Kraus2013"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">my_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">pathToFile</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span> <span class="co">#number of individuals</span>
<span class="co">#&gt; [1] 695</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">)</span> <span class="co">#number of flyways</span>
<span class="co">#&gt; [1] 10</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span> <span class="co">#number of individuals per flyway</span>
<span class="co">#&gt; my_groups</span>
<span class="co">#&gt;                        AsiaCentral                           AsiaEast                   AsiaSouthWestern </span>
<span class="co">#&gt;                                 51                                 12                                 15 </span>
<span class="co">#&gt; EuropeBlackSeaEasternMediterranean                 EuropeNorthWestern         EuropeWesternMediterranean </span>
<span class="co">#&gt;                                 21                                147                                360 </span>
<span class="co">#&gt;                          Greenland               NorthAmericaAtlantic                NorthAmericaCentral </span>
<span class="co">#&gt;                                 29                                 16                                 22 </span>
<span class="co">#&gt;                NorthAmericaPacific </span>
<span class="co">#&gt;                                 22</span></code></pre></div>
<p>Number of populations (not needed for analysis hereafter):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_pops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">pathToFile</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">my_pops</span><span class="op">)</span><span class="op">)</span> <span class="co">#number of populations</span>
<span class="co">#&gt; [1] 55</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">my_pops</span><span class="op">)</span> <span class="co">#number of individuals per population</span>
<span class="co">#&gt; my_pops</span>
<span class="co">#&gt;  ATHO  CACO  CAEK  CAJC  CALM  CARM  CASL  CNLI  CYLA  DEWU  EETA  FIOU  FOTO  FRAL  FRMV  GLIS  GLNU  GREV IR361 </span>
<span class="co">#&gt;    25     4     4     1     7    20     2     5     5    27    22    19    24    10    32     9    20    16     1 </span>
<span class="co">#&gt; IR362 IR366  IR37 IR371 IR375 IR377   IR4 IR444 IR448 IR449 IR452 IR454 IR513 IR514  ISHV  LTVE  NLFR  NOBE  NOSS </span>
<span class="co">#&gt;     1     1     1     1     1     1     1     1     1     1     1     1     1     1     4    17    32    32    16 </span>
<span class="co">#&gt;  PKHA  PTDJ  RUIV  RUKH  RULE  RUNO  RUOM  RUTO  RUTV  RUVL  RUVO  RUYA  SEOB  SILJ  UADU  UALV  USMF </span>
<span class="co">#&gt;     7    32    27     7    31     8    12    32    19    29    31    25    20    19     3     3    22</span></code></pre></div>
<p>Code per individual (not needed for analysis hereafter):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_indv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">pathToFile</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>SNP names (not needed for analysis hereafter):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_snps_Kraus2013"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">my_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">pathToFile</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">my_snps</span><span class="op">)</span> <span class="co"># number of snps</span>
<span class="co">#&gt; [1] 364</span></code></pre></div>
</div>
<div id="run-smart_pca" class="section level2">
<h2 class="hasAnchor">
<a href="#run-smart_pca" class="anchor"></a>Run <em>smart_pca</em>
</h2>
<p>Run PCA with truncated SVD (PCA 1 x PCA 2 axes) and assign results to object pcaR (missing values imputed with means, SNPs scaled to control genetic drift):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_genotype_Kraus2012"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">pcaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_pca.html">smart_pca</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, missing_impute <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></code></pre></div>
<p>pcaR is a list with 3 elements:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">pcaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pcaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "pca.snp_loadings"       "pca.eigenvalues"        "pca.sample_coordinates"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">pcaR</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ pca.snp_loadings      :'data.frame':  363 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ PC1: num [1:363] -0.00485 0.01722 0.01528 -0.03058 -0.01299 ...</span>
<span class="co">#&gt;   ..$ PC2: num [1:363] 0.12022 0.02884 0.10416 0.03784 -0.00292 ...</span>
<span class="co">#&gt;  $ pca.eigenvalues       : num [1:3, 1:2] 14.24 1.93 1.93 11.89 1.61 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:3] "observed eigenvalues" "variance explained" "cumulative variance explained"</span>
<span class="co">#&gt;   .. ..$ : chr [1:2] "PC1" "PC2"</span>
<span class="co">#&gt;  $ pca.sample_coordinates:'data.frame':  695 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ Group: chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   ..$ Class: chr [1:695] "PCA" "PCA" "PCA" "PCA" ...</span>
<span class="co">#&gt;   ..$ PC1  : num [1:695] 1.194 1.654 -1.152 1.489 -0.976 ...</span>
<span class="co">#&gt;   ..$ PC2  : num [1:695] 2.42 2.86 3.31 3.9 3.21 ...</span></code></pre></div>
<p>Assign statistical results to objects pcaR_eigen, pcaR_load and pcaR_coord:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pcaR_eigen</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.eigenvalues</span> <span class="co"># extract eigenvalues (PCA1 and PC2 axes explain 3.5% variation in SNP variation across individuals)</span>
<span class="va">pcaR_load</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.snp_loadings</span> <span class="co"># extract principal coefficients (high SNP loadings indicate loci with stronger variation across individuals)</span>
<span class="va">pcaR_coord</span> <span class="op">&lt;-</span> <span class="va">pcaR</span><span class="op">$</span><span class="va">pca.sample_coordinates</span> <span class="co"># extract principal components (position of individuals in PCA space used to generate the ordination)</span></code></pre></div>
<p>Plot PCA:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pcaR</span><span class="op">$</span><span class="va">pca.sample_coordinates</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>,<span class="st">"PC2"</span><span class="op">)</span><span class="op">]</span>, cex <span class="op">=</span> <span class="fl">1.5</span>, 
     bg <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">21</span>, col <span class="op">=</span> <span class="st">"black"</span>, main <span class="op">=</span> <span class="st">"mallard genotype smartpca"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">my_groups</span><span class="op">)</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fl">1</span>, pch <span class="op">=</span> <span class="fl">21</span>, 
       pt.cex <span class="op">=</span> <span class="fl">1.0</span>, col <span class="op">=</span> <span class="st">"black"</span>, pt.bg <span class="op">=</span> <span class="va">cols</span>, text.col <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="figure/pca_plot_mallard-1.png" title="plot of chunk pca_plot_mallard" alt="plot of chunk pca_plot_mallard" style="display: block; margin: auto;"></p>
<p>Greenland individuals cluster in one of the corners of the ordination, supporting a distinct SNP composition relative to the remaining flyways.</p>
</div>
<div id="run-smart_permanova" class="section level2">
<h2 class="hasAnchor">
<a href="#run-smart_permanova" class="anchor"></a>Run <em>smart_permanova</em>
</h2>
<p>Run PERMANOVA test (group location in PCA1 x PCA2 space) and assign results to object permanovaR (missing values imputed with means, SNPs scaled to control genetic drift). Notice that pairwise tests increase computing time considerably as there are 45 pairwise comparisons to make for 10 flyways, each calculating a p value based on 10,000 permutations of the data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_genotype_Kraus2012"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">permanovaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_permanova.html">smart_permanova</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, 
                              target_space <span class="op">=</span> <span class="st">"pca"</span>, missing_impute <span class="op">=</span> <span class="st">"mean"</span>, pairwise <span class="op">=</span> <span class="st">"TRUE"</span><span class="op">)</span></code></pre></div>
<p>permanovaR is a list with 5 elements:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">permanovaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">permanovaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "permanova.samples"             "permanova.global_test"         "permanova.pairwise_test"      </span>
<span class="co">#&gt; [4] "permanova.pairwise_correction" "permanova.permutation_number"  "permanova.permutation_seed"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">permanovaR</span><span class="op">)</span>
<span class="co">#&gt; List of 6</span>
<span class="co">#&gt;  $ permanova.samples            :'data.frame':   695 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ Group: chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   ..$ Class: chr [1:695] "PERMANOVA" "PERMANOVA" "PERMANOVA" "PERMANOVA" ...</span>
<span class="co">#&gt;  $ permanova.global_test        :Classes 'anova' and 'data.frame':   3 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ Df       : num [1:3] 9 685 694</span>
<span class="co">#&gt;   ..$ SumsOfSqs: num [1:3] 6982 11151 18133</span>
<span class="co">#&gt;   ..$ MeanSqs  : num [1:3] 775.7 16.3 NA</span>
<span class="co">#&gt;   ..$ F.Model  : num [1:3] 47.7 NA NA</span>
<span class="co">#&gt;   ..$ R2       : num [1:3] 0.385 0.615 1</span>
<span class="co">#&gt;   ..$ Pr(&gt;F)   : num [1:3] 1e-04 NA NA</span>
<span class="co">#&gt;  $ permanova.pairwise_test      :'data.frame':   45 obs. of  5 variables:</span>
<span class="co">#&gt;   ..$ GroupPair        : chr [1:45] "AsiaCentral-AsiaEast" "AsiaCentral-AsiaSouthWestern" "AsiaCentral-EuropeBlackSeaEasternMediterranean" "AsiaCentral-EuropeNorthWestern" ...</span>
<span class="co">#&gt;   ..$ F.Model          : num [1:45] 1.366 4.547 0.672 34.286 4.738 ...</span>
<span class="co">#&gt;   ..$ R2               : num [1:45] 0.0219 0.06633 0.00951 0.14888 0.01145 ...</span>
<span class="co">#&gt;   ..$ P.value          : num [1:45] 0.2505 0.0414 0.5117 0.0001 0.0211 ...</span>
<span class="co">#&gt;   ..$ P.value.corrected: num [1:45] 1 0.8694 1 0.0045 0.5408 ...</span>
<span class="co">#&gt;  $ permanova.pairwise_correction: chr "holm"</span>
<span class="co">#&gt;  $ permanova.permutation_number : num 9999</span>
<span class="co">#&gt;  $ permanova.permutation_seed   : num 1</span></code></pre></div>
<p>Assign sample summary to object permP:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">permP</span> <span class="op">&lt;-</span> <span class="va">permanovaR</span><span class="op">$</span><span class="va">permanova.samples</span></code></pre></div>
<p>Show PERMANOVA tables (global and pairwise):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">permanovaR</span><span class="op">$</span><span class="va">permanova.global_test</span>
<span class="co">#&gt;            Df SumsOfSqs MeanSqs F.Model      R2 Pr(&gt;F)    </span>
<span class="co">#&gt; group       9    6981.7  775.74  47.653 0.38503  1e-04 ***</span>
<span class="co">#&gt; Residuals 685   11151.1   16.28         0.61497           </span>
<span class="co">#&gt; Total     694   18132.8                 1.00000           </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></code></pre></div>
<p>For the mallard dataset, the p value is 1e-04. As with other frequentist tests, p values should be interpreted as the probability of the observed differences if the null hypothesis of no differences between groups is true. The lower the p value, the weaker the support for the null hypothesis.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">permanovaR</span><span class="op">$</span><span class="va">permanova.pairwise_test</span><span class="op">)</span>
<span class="co">#&gt;                                        GroupPair     F.Model          R2 P.value P.value.corrected</span>
<span class="co">#&gt; 1                           AsiaCentral-AsiaEast   1.3655903 0.021896534  0.2505            1.0000</span>
<span class="co">#&gt; 2                   AsiaCentral-AsiaSouthWestern   4.5468589 0.066332126  0.0414            0.8694</span>
<span class="co">#&gt; 3 AsiaCentral-EuropeBlackSeaEasternMediterranean   0.6718692 0.009506883  0.5117            1.0000</span>
<span class="co">#&gt; 4                 AsiaCentral-EuropeNorthWestern  34.2856122 0.148882997  0.0001            0.0045</span>
<span class="co">#&gt; 5         AsiaCentral-EuropeWesternMediterranean   4.7384432 0.011452751  0.0211            0.5408</span>
<span class="co">#&gt; 6                          AsiaCentral-Greenland 314.3271374 0.801186325  0.0001            0.0045</span></code></pre></div>
<p>The lowest p values (resulting from pairwise comparisons) consistently occur between the Greenland and the remaining flyways, supporting a unique SNP composition mostly in Greenland mallards.</p>
</div>
<div id="run-smart_permdisp" class="section level2">
<h2 class="hasAnchor">
<a href="#run-smart_permdisp" class="anchor"></a>Run <em>smart_permdisp</em>
</h2>
<p>Run PERMDISP test (group dispersion in PCA1 x PCA2 space) and assign results to object permdispR (missing values imputed with means, SNPs scaled to control genetic drift). Heteroscededasticity tests in combination with ANOVA tests tell whether the ANOVA F statistic is driven by mean and/or varinance differences among groups in a univariate context. Location and dispersion (multivariate context) are analogous with mean and variance in a univariate context. As the number per individuals per flyway differ a great deal among flyways, PERMDISP is run to control for sample-size bias (samplesize_bias = TRUE).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_genotype_Kraus2012"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">permdispR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_permdisp.html">smart_permdisp</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, 
                            target_space <span class="op">=</span> <span class="st">"pca"</span>, missing_impute <span class="op">=</span> <span class="st">"mean"</span>, pairwise <span class="op">=</span> <span class="st">"TRUE"</span>, samplesize_bias <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>permdispR is a list with 7 elements:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">permdispR</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">permdispR</span><span class="op">)</span>
<span class="co">#&gt; [1] "permdisp.samples"             "permdisp_bias"                "permdisp.group_location"     </span>
<span class="co">#&gt; [4] "permdisp.global_test"         "permdisp.pairwise_test"       "permdisp.pairwise_correction"</span>
<span class="co">#&gt; [7] "permdisp.permutation_number"  "permdisp.permutation_seed"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">permdispR</span><span class="op">)</span>
<span class="co">#&gt; List of 8</span>
<span class="co">#&gt;  $ permdisp.samples            :'data.frame':    695 obs. of  3 variables:</span>
<span class="co">#&gt;   ..$ Group            : chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   ..$ Class            : chr [1:695] "PERMDISP" "PERMDISP" "PERMDISP" "PERMDISP" ...</span>
<span class="co">#&gt;   ..$ Sample_dispersion: num [1:695] 1.91 2.53 2.85 3.36 2.68 ...</span>
<span class="co">#&gt;  $ permdisp_bias               : chr "Dispersion adjusted to number of samples per group"</span>
<span class="co">#&gt;  $ permdisp.group_location     : num [1:10, 1:2] 0.956 1.739 0.963 0.71 -1.464 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:10] "AsiaCentral" "AsiaEast" "AsiaSouthWestern" "EuropeBlackSeaEasternMediterranean" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:2] "PCoA1" "PCoA2"</span>
<span class="co">#&gt;  $ permdisp.global_test        :'data.frame':    2 obs. of  5 variables:</span>
<span class="co">#&gt;   ..$ Df     : int [1:2] 9 685</span>
<span class="co">#&gt;   ..$ Sum Sq : num [1:2] 258 6128</span>
<span class="co">#&gt;   ..$ Mean Sq: num [1:2] 28.7 8.95</span>
<span class="co">#&gt;   ..$ F      : num [1:2] 3.21 NA</span>
<span class="co">#&gt;   ..$ Pr(&gt;F) : num [1:2] 0.0073 NA</span>
<span class="co">#&gt;  $ permdisp.pairwise_test      :'data.frame':    45 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ GroupPair        : chr [1:45] "AsiaCentral-AsiaEast" "AsiaCentral-AsiaSouthWestern" "AsiaCentral-EuropeBlackSeaEasternMediterranean" "AsiaCentral-EuropeNorthWestern" ...</span>
<span class="co">#&gt;   ..$ F.Model          : num [1:45] 0.00367 5.20981 0.13897 20.46506 1.93196 ...</span>
<span class="co">#&gt;   ..$ P.value          : num [1:45] 0.9527 0.0293 0.7053 0.0001 0.1502 ...</span>
<span class="co">#&gt;   ..$ P.value.corrected: num [1:45] 1 1 1 0.0045 1 1 1 1 1 1 ...</span>
<span class="co">#&gt;  $ permdisp.pairwise_correction: chr "holm"</span>
<span class="co">#&gt;  $ permdisp.permutation_number : num 9999</span>
<span class="co">#&gt;  $ permdisp.permutation_seed   : num 1</span></code></pre></div>
<p>Assign sample summary to object permD, where column Sample_dispersion column show dispersion of individuals relative to their flyway:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">permD</span> <span class="op">&lt;-</span> <span class="va">permdispR</span><span class="op">$</span><span class="va">permdisp.samples</span></code></pre></div>
<p>Show PERMDISP tables (global and pairwise):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">permdispR</span><span class="op">$</span><span class="va">permdisp.global_test</span>
<span class="co">#&gt;            Df    Sum Sq   Mean Sq       F Pr(&gt;F)</span>
<span class="co">#&gt; Groups      9  258.2757 28.697299 3.20762 0.0073</span>
<span class="co">#&gt; Residuals 685 6128.4218  8.946601      NA     NA</span></code></pre></div>
<p>For the mallard dataset, the p value is 0.0073:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">permdispR</span><span class="op">$</span><span class="va">permdisp.pairwise_test</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    45 obs. of  4 variables:</span>
<span class="co">#&gt;  $ GroupPair        : chr  "AsiaCentral-AsiaEast" "AsiaCentral-AsiaSouthWestern" "AsiaCentral-EuropeBlackSeaEasternMediterranean" "AsiaCentral-EuropeNorthWestern" ...</span>
<span class="co">#&gt;  $ F.Model          : num  0.00367 5.20981 0.13897 20.46506 1.93196 ...</span>
<span class="co">#&gt;  $ P.value          : num  0.9527 0.0293 0.7053 0.0001 0.1502 ...</span>
<span class="co">#&gt;  $ P.value.corrected: num  1 1 1 0.0045 1 1 1 1 1 1 ...</span></code></pre></div>
<p>Most PERMDISP pairwise tests show relatively high p values (i.e., high probability of the observed differences in dispersion if the null hypothesis of no dispersion differences among groups is true), indicating that PERMANOVA tests mainly captured differences in location. The lowest p values for the PERMDISP pairwise tests among Eurasian flyways occur for the Europe North Western (ENW) flyway versus the other flyways as seen in the ordination plot (i.e., ENW individuals are widely spread over both the PCA1 and PCA2 axes).</p>
</div>
<div id="run-smart_mva" class="section level2">
<h2 class="hasAnchor">
<a href="#run-smart_mva" class="anchor"></a>Run <em>smart_mva</em>
</h2>
<p>Run PCA, and PERMANOVA and PERMDISP tests (group location and dispersion in PCA1 x PCA2 space), and assign results to object mvaR. No pairwise comparisons are applied (default: pairwise = “FALSE”), so computation will be relatively fast. This is a wrapper function running in one single job the three other functions of the package (smart_pca, smart_permanova, smart_permdisp).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathToFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mallard_genotype_Kraus2012"</span>, package <span class="op">=</span> <span class="st">"smartsnp"</span><span class="op">)</span>
<span class="va">mvaR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_mva.html">smart_mva</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="va">pathToFile</span>, sample_group <span class="op">=</span> <span class="va">my_groups</span>, 
                            target_space <span class="op">=</span> <span class="st">"pca"</span>, missing_impute <span class="op">=</span> <span class="st">"mean"</span>, samplesize_bias <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>mvaR is a list with three elements (data, pca, test):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">)</span>
<span class="co">#&gt; [1] "data" "pca"  "test"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ data: num [1:363, 1:695] -1.13 1.31 2.05 -1.33 -0.81 ...</span>
<span class="co">#&gt;  $ pca :List of 3</span>
<span class="co">#&gt;   ..$ pca.snp_loadings      :'data.frame':   363 obs. of  2 variables:</span>
<span class="co">#&gt;   .. ..$ PC1: num [1:363] -0.00485 0.01722 0.01528 -0.03058 -0.01299 ...</span>
<span class="co">#&gt;   .. ..$ PC2: num [1:363] 0.12022 0.02884 0.10416 0.03784 -0.00292 ...</span>
<span class="co">#&gt;   ..$ pca.eigenvalues       : num [1:3, 1:2] 14.24 1.93 1.93 11.89 1.61 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:3] "observed eigenvalues" "variance explained" "cumulative variance explained"</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:2] "PC1" "PC2"</span>
<span class="co">#&gt;   ..$ pca.sample_coordinates:'data.frame':   695 obs. of  4 variables:</span>
<span class="co">#&gt;   .. ..$ Group: chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   .. ..$ Class: chr [1:695] "PCA" "PCA" "PCA" "PCA" ...</span>
<span class="co">#&gt;   .. ..$ PC1  : num [1:695] 1.194 1.654 -1.152 1.489 -0.976 ...</span>
<span class="co">#&gt;   .. ..$ PC2  : num [1:695] 2.42 2.86 3.31 3.9 3.21 ...</span>
<span class="co">#&gt;  $ test:List of 10</span>
<span class="co">#&gt;   ..$ test_samples            :'data.frame': 695 obs. of  3 variables:</span>
<span class="co">#&gt;   .. ..$ Group            : chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   .. ..$ Class            : chr [1:695] "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" ...</span>
<span class="co">#&gt;   .. ..$ Sample_dispersion: num [1:695] 1.91 2.53 2.85 3.36 2.68 ...</span>
<span class="co">#&gt;   ..$ permanova.global_test   :Classes 'anova' and 'data.frame': 3 obs. of  6 variables:</span>
<span class="co">#&gt;   .. ..$ Df       : num [1:3] 9 685 694</span>
<span class="co">#&gt;   .. ..$ SumsOfSqs: num [1:3] 6982 11151 18133</span>
<span class="co">#&gt;   .. ..$ MeanSqs  : num [1:3] 775.7 16.3 NA</span>
<span class="co">#&gt;   .. ..$ F.Model  : num [1:3] 47.7 NA NA</span>
<span class="co">#&gt;   .. ..$ R2       : num [1:3] 0.385 0.615 1</span>
<span class="co">#&gt;   .. ..$ Pr(&gt;F)   : num [1:3] 1e-04 NA NA</span>
<span class="co">#&gt;   ..$ permanova.pairwise_test : chr "No PERMANOVA pairwise tests implemented"</span>
<span class="co">#&gt;   ..$ permdisp.global_test    :'data.frame': 2 obs. of  5 variables:</span>
<span class="co">#&gt;   .. ..$ Df     : int [1:2] 9 685</span>
<span class="co">#&gt;   .. ..$ Sum Sq : num [1:2] 258 6128</span>
<span class="co">#&gt;   .. ..$ Mean Sq: num [1:2] 28.7 8.95</span>
<span class="co">#&gt;   .. ..$ F      : num [1:2] 3.21 NA</span>
<span class="co">#&gt;   .. ..$ Pr(&gt;F) : num [1:2] 0.0095 NA</span>
<span class="co">#&gt;   ..$ permdisp.pairwise_test  : chr "No PERMDISP pairwise tests implemented"</span>
<span class="co">#&gt;   ..$ permdisp.bias           : chr "Dispersion adjusted to number of samples per group"</span>
<span class="co">#&gt;   ..$ permdisp.group_location : num [1:10, 1:2] 0.956 1.739 0.963 0.71 -1.464 ...</span>
<span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:10] "AsiaCentral" "AsiaEast" "AsiaSouthWestern" "EuropeBlackSeaEasternMediterranean" ...</span>
<span class="co">#&gt;   .. .. ..$ : chr [1:2] "PCoA1" "PCoA2"</span>
<span class="co">#&gt;   ..$ test.pairwise_correction: chr "holm"</span>
<span class="co">#&gt;   ..$ test.permutation_number : num 9999</span>
<span class="co">#&gt;   ..$ test.permutation_seed   : num 1</span></code></pre></div>
<p>Element 1 = scaled dataset (none, covariance, correlation, drift) in a matrix and array (rows = SNPs, columns = samples):</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; [1] "matrix" "array"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; [1] 363 695</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt;  num [1:363, 1:695] -1.13 1.31 2.05 -1.33 -0.81 ...</span></code></pre></div>
<p>Element 2 = PCA results in a list:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">)</span>
<span class="co">#&gt; [1] "pca.snp_loadings"       "pca.eigenvalues"        "pca.sample_coordinates"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ pca.snp_loadings      :'data.frame':  363 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ PC1: num [1:363] -0.00485 0.01722 0.01528 -0.03058 -0.01299 ...</span>
<span class="co">#&gt;   ..$ PC2: num [1:363] 0.12022 0.02884 0.10416 0.03784 -0.00292 ...</span>
<span class="co">#&gt;  $ pca.eigenvalues       : num [1:3, 1:2] 14.24 1.93 1.93 11.89 1.61 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:3] "observed eigenvalues" "variance explained" "cumulative variance explained"</span>
<span class="co">#&gt;   .. ..$ : chr [1:2] "PC1" "PC2"</span>
<span class="co">#&gt;  $ pca.sample_coordinates:'data.frame':  695 obs. of  4 variables:</span>
<span class="co">#&gt;   ..$ Group: chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   ..$ Class: chr [1:695] "PCA" "PCA" "PCA" "PCA" ...</span>
<span class="co">#&gt;   ..$ PC1  : num [1:695] 1.194 1.654 -1.152 1.489 -0.976 ...</span>
<span class="co">#&gt;   ..$ PC2  : num [1:695] 2.42 2.86 3.31 3.9 3.21 ...</span></code></pre></div>
<p>Show PCA results:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.eigenvalues</span><span class="op">)</span> <span class="co">#extract eigenvalues</span>
<span class="co">#&gt;                                     PC1       PC2</span>
<span class="co">#&gt; observed eigenvalues          14.235721 11.892240</span>
<span class="co">#&gt; variance explained             1.925574  1.608587</span>
<span class="co">#&gt; cumulative variance explained  1.925574  3.534161</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.sample_coordinates</span><span class="op">)</span> <span class="co">#extract coordinates of individuals in PCA1 x PCA2 space</span>
<span class="co">#&gt;                        Group Class        PC1      PC2</span>
<span class="co">#&gt; 1 EuropeWesternMediterranean   PCA  1.1936564 2.420538</span>
<span class="co">#&gt; 2 EuropeWesternMediterranean   PCA  1.6540335 2.857820</span>
<span class="co">#&gt; 3 EuropeWesternMediterranean   PCA -1.1522899 3.311944</span>
<span class="co">#&gt; 4 EuropeWesternMediterranean   PCA  1.4886777 3.895509</span>
<span class="co">#&gt; 5 EuropeWesternMediterranean   PCA -0.9762505 3.214546</span>
<span class="co">#&gt; 6 EuropeWesternMediterranean   PCA -0.3518813 2.273576</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">pca</span><span class="op">$</span><span class="va">pca.snp_loadings</span><span class="op">)</span> <span class="co">#extract SNP loadings</span>
<span class="co">#&gt;            PC1          PC2</span>
<span class="co">#&gt; 1 -0.004845369  0.120219092</span>
<span class="co">#&gt; 2  0.017218374  0.028841552</span>
<span class="co">#&gt; 3  0.015280677  0.104157428</span>
<span class="co">#&gt; 4 -0.030582459  0.037843697</span>
<span class="co">#&gt; 5 -0.012987416 -0.002920648</span>
<span class="co">#&gt; 6 -0.098992766 -0.028586880</span></code></pre></div>
<p>Element 3 = PERMANOVA and PERMDISP results in a list:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">)</span>
<span class="co">#&gt; [1] "list"</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">)</span>
<span class="co">#&gt;  [1] "test_samples"             "permanova.global_test"    "permanova.pairwise_test"  "permdisp.global_test"    </span>
<span class="co">#&gt;  [5] "permdisp.pairwise_test"   "permdisp.bias"            "permdisp.group_location"  "test.pairwise_correction"</span>
<span class="co">#&gt;  [9] "test.permutation_number"  "test.permutation_seed"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">)</span>
<span class="co">#&gt; List of 10</span>
<span class="co">#&gt;  $ test_samples            :'data.frame':    695 obs. of  3 variables:</span>
<span class="co">#&gt;   ..$ Group            : chr [1:695] "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" "EuropeWesternMediterranean" ...</span>
<span class="co">#&gt;   ..$ Class            : chr [1:695] "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" "PERMANOVA/PERMDISP" ...</span>
<span class="co">#&gt;   ..$ Sample_dispersion: num [1:695] 1.91 2.53 2.85 3.36 2.68 ...</span>
<span class="co">#&gt;  $ permanova.global_test   :Classes 'anova' and 'data.frame':    3 obs. of  6 variables:</span>
<span class="co">#&gt;   ..$ Df       : num [1:3] 9 685 694</span>
<span class="co">#&gt;   ..$ SumsOfSqs: num [1:3] 6982 11151 18133</span>
<span class="co">#&gt;   ..$ MeanSqs  : num [1:3] 775.7 16.3 NA</span>
<span class="co">#&gt;   ..$ F.Model  : num [1:3] 47.7 NA NA</span>
<span class="co">#&gt;   ..$ R2       : num [1:3] 0.385 0.615 1</span>
<span class="co">#&gt;   ..$ Pr(&gt;F)   : num [1:3] 1e-04 NA NA</span>
<span class="co">#&gt;  $ permanova.pairwise_test : chr "No PERMANOVA pairwise tests implemented"</span>
<span class="co">#&gt;  $ permdisp.global_test    :'data.frame':    2 obs. of  5 variables:</span>
<span class="co">#&gt;   ..$ Df     : int [1:2] 9 685</span>
<span class="co">#&gt;   ..$ Sum Sq : num [1:2] 258 6128</span>
<span class="co">#&gt;   ..$ Mean Sq: num [1:2] 28.7 8.95</span>
<span class="co">#&gt;   ..$ F      : num [1:2] 3.21 NA</span>
<span class="co">#&gt;   ..$ Pr(&gt;F) : num [1:2] 0.0095 NA</span>
<span class="co">#&gt;  $ permdisp.pairwise_test  : chr "No PERMDISP pairwise tests implemented"</span>
<span class="co">#&gt;  $ permdisp.bias           : chr "Dispersion adjusted to number of samples per group"</span>
<span class="co">#&gt;  $ permdisp.group_location : num [1:10, 1:2] 0.956 1.739 0.963 0.71 -1.464 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:10] "AsiaCentral" "AsiaEast" "AsiaSouthWestern" "EuropeBlackSeaEasternMediterranean" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:2] "PCoA1" "PCoA2"</span>
<span class="co">#&gt;  $ test.pairwise_correction: chr "holm"</span>
<span class="co">#&gt;  $ test.permutation_number : num 9999</span>
<span class="co">#&gt;  $ test.permutation_seed   : num 1</span></code></pre></div>
<p>Multiple-testing correction applied:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test.pairwise_correction</span>
<span class="co">#&gt; [1] "holm"</span></code></pre></div>
<p>Number of permutations to estimate p value:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test.permutation_number</span>
<span class="co">#&gt; [1] 9999</span></code></pre></div>
<p>Seed for random generator:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test.permutation_seed</span>
<span class="co">#&gt; [1] 1</span></code></pre></div>
<p>Summary of samples:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test_samples</span><span class="op">)</span>
<span class="co">#&gt;                        Group              Class Sample_dispersion</span>
<span class="co">#&gt; 1 EuropeWesternMediterranean PERMANOVA/PERMDISP          1.905208</span>
<span class="co">#&gt; 2 EuropeWesternMediterranean PERMANOVA/PERMDISP          2.525640</span>
<span class="co">#&gt; 3 EuropeWesternMediterranean PERMANOVA/PERMDISP          2.846770</span>
<span class="co">#&gt; 4 EuropeWesternMediterranean PERMANOVA/PERMDISP          3.357369</span>
<span class="co">#&gt; 5 EuropeWesternMediterranean PERMANOVA/PERMDISP          2.679859</span>
<span class="co">#&gt; 6 EuropeWesternMediterranean PERMANOVA/PERMDISP          1.566851</span></code></pre></div>
<p>Show PERMANOVA table:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permanova.global_test</span> <span class="co">#global test</span>
<span class="co">#&gt;            Df SumsOfSqs MeanSqs F.Model      R2 Pr(&gt;F)    </span>
<span class="co">#&gt; group       9    6981.7  775.74  47.653 0.38503  1e-04 ***</span>
<span class="co">#&gt; Residuals 685   11151.1   16.28         0.61497           </span>
<span class="co">#&gt; Total     694   18132.8                 1.00000           </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permanova.pairwise_test</span> <span class="co">#pairwise tests</span>
<span class="co">#&gt; [1] "No PERMANOVA pairwise tests implemented"</span></code></pre></div>
<p>Show PERMDISP table:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permdisp.global_test</span> <span class="co">#global test</span>
<span class="co">#&gt;            Df    Sum Sq   Mean Sq       F Pr(&gt;F)</span>
<span class="co">#&gt; Groups      9  258.2757 28.697299 3.20762 0.0095</span>
<span class="co">#&gt; Residuals 685 6128.4218  8.946601      NA     NA</span>
<span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permdisp.pairwise_test</span> <span class="co">#pairwise tests</span>
<span class="co">#&gt; NULL</span></code></pre></div>
<p>Sample-size correction applied:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permdisp.bias</span>
<span class="co">#&gt; [1] "Dispersion adjusted to number of samples per group"</span></code></pre></div>
<p>Location of flyways in ordination:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mvaR</span><span class="op">$</span><span class="va">test</span><span class="op">$</span><span class="va">permdisp.group_location</span>
<span class="co">#&gt;                                          PCoA1      PCoA2</span>
<span class="co">#&gt; AsiaCentral                         0.95606737 -1.0223659</span>
<span class="co">#&gt; AsiaEast                            1.73944124 -1.4799962</span>
<span class="co">#&gt; AsiaSouthWestern                    0.96254119 -1.2005153</span>
<span class="co">#&gt; EuropeBlackSeaEasternMediterranean  0.70978842 -1.0906961</span>
<span class="co">#&gt; EuropeNorthWestern                 -1.46377155  0.5649683</span>
<span class="co">#&gt; EuropeWesternMediterranean          0.18849627 -0.8051776</span>
<span class="co">#&gt; Greenland                           9.37848059 10.5513341</span>
<span class="co">#&gt; NorthAmericaAtlantic               -0.03785058 -0.3764883</span>
<span class="co">#&gt; NorthAmericaCentral                 1.61458028 -2.6342789</span>
<span class="co">#&gt; NorthAmericaPacific                 1.38064146 -1.7098544</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
