<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Smart Multivariate Analyses (wrapper of PCA, PERMANOVA and PERMDISP) — smart_mva • smartsnp</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Smart Multivariate Analyses (wrapper of PCA, PERMANOVA and PERMDISP) — smart_mva" />
<meta property="og:description" content="Computes Principal Component Analysis (PCA) for variable x sample genotype data, such as Single Nucleotide Polymorphisms (SNP), in combination with Permutational Multivariate Analysis of Variance (PERMANOVA) and Permutational Multivariate Analysis of Dispersion (PERMDISP).
A wrapper of functions smart_pca, smart_permanova and smart_permdisp.
Genetic markers such as SNPs can be scaled by centering, z-scores and genetic drift-based dispersion.
The latter follows the SMARTPCA implementation of Patterson, Price and Reich (2006).
Optimized to run fast computation for big datasets." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="../articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="../articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Smart Multivariate Analyses (wrapper of PCA, PERMANOVA and PERMDISP)</h1>
    
    <div class="hidden name"><code>smart_mva.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Computes Principal Component Analysis (PCA) for variable x sample genotype data, such as Single Nucleotide Polymorphisms (SNP), in combination with Permutational Multivariate Analysis of Variance (PERMANOVA) and Permutational Multivariate Analysis of Dispersion (PERMDISP).
A wrapper of functions <code>smart_pca</code>, <code>smart_permanova</code> and <code>smart_permdisp</code>.
Genetic markers such as SNPs can be scaled by <code>centering</code>, z-scores and genetic drift-based dispersion.
The latter follows the SMARTPCA implementation of Patterson, Price and Reich (2006).
Optimized to run fast computation for big datasets.</p>
    </div>


    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>snp_data</th>
      <td><p>snp_data</p></td>
    </tr>
    <tr>
      <th>packed_data</th>
      <td><p>Logical value for <code>EIGENSTRAT</code>, irrelevant for text data.
Default <code>packed_data = FALSE</code> assumes uncompressed <code>EIGENSTRAT</code>.
<code>packed_data = TRUE</code> for compressed or binary <code>EIGENSTRAT</code> (<code>PACKENDANCESTRYMAP</code>).</p></td>
    </tr>
    <tr>
      <th>sample_group</th>
      <td><p>Character or numeric vector assigning samples to groups.
Coerced to factor.</p></td>
    </tr>
    <tr>
      <th>sample_remove</th>
      <td><p>Logical <code>FALSE</code> or numeric vector indicating column numbers (samples) to be removed from computations.
Default <code>sample_remove =  FALSE</code> keeps all samples.</p></td>
    </tr>
    <tr>
      <th>snp_remove</th>
      <td><p>Logical <code>FALSE</code> or numeric vector indicating row numbers (SNPs) to be removed from computations.
Default <code>snp_remove =  FALSE</code> keeps all SNPs.
See details.</p></td>
    </tr>
    <tr>
      <th>pca</th>
      <td><p>Logical indicating if PCA is computed.
Default <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>permanova</th>
      <td><p>Logical indicating if PERMANOVA is computed.
Default <code>TRUE</code></p></td>
    </tr>
    <tr>
      <th>permdisp</th>
      <td><p>Logical indicating if PERMDISP is computed.
Default <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>missing_value</th>
      <td><p>Number <code>9</code> or string <code>NA</code> indicating missing value.
Default <code>missing_value = 9</code> as in <code>EIGENSTRAT</code>.
If no missing values present, no effect on computation.</p></td>
    </tr>
    <tr>
      <th>missing_impute</th>
      <td><p>String handling missing values.
Default <code>missing_impute = "mean"</code> replaces missing values of each SNP by mean of non-missing values across samples.
<code>missing_impute = "remove"</code> removes SNPs with at least one missing value.
If no missing values present, no effect on computation.</p></td>
    </tr>
    <tr>
      <th>scaling</th>
      <td><p>String. Default <code>scaling = "drift"</code> scales SNPs to control for expected allele frequency dispersion caused by genetic drift (SMARTPCA).
<code>scaling = "center"</code> for <code>centering</code> (covariance-based PCA).
<code>scaling = "sd"</code> for <code>centered</code> SNPs divided by standard deviation (correlation-based PCA).
<code>scaling = "none"</code> for no scaling.
See details.</p></td>
    </tr>
    <tr>
      <th>program_svd</th>
      <td><p>String indicating R package computing single value decomposition (SVD).
Default <code>program_svd = "Rspectra"</code> for <code><a href='https://rdrr.io/pkg/RSpectra/man/svds.html'>svds</a></code>.
<code>program_svd = "bootSVD"</code> for <code><a href='https://rdrr.io/pkg/bootSVD/man/fastSVD.html'>fastSVD</a></code>.
See details.</p></td>
    </tr>
    <tr>
      <th>sample_project</th>
      <td><p>Numeric vector indicating column numbers (ancient samples) projected onto (modern) PCA space.
Default <code>sample_project =  FALSE</code> implements no projection.
See details.</p></td>
    </tr>
    <tr>
      <th>pc_project</th>
      <td><p>Numeric vector indicating the ranks of the PCA axes ancient samples are projected onto. Default <code>pc_ancient = c(1, 2)</code> for PCA axes 1 and 2.
If <code>program_svd = "RSpectra"</code>, <code><a href='https://rdrr.io/r/base/length.html'>length(pc_ancient)</a></code> must be smaller than or equal to <code>pc_axes</code>.
No effect on computation, if no ancient samples present.</p></td>
    </tr>
    <tr>
      <th>sample_distance</th>
      <td><p>Type of inter-sample proximity computed (distance, similarity, dissimilarity).
Default is <code>Euclidean distance</code>.
See details.</p></td>
    </tr>
    <tr>
      <th>program_distance</th>
      <td><p>A string value indicating R package to estimate proximities between pairs of samples.
Default <code>program_distance = "Rfast"</code> uses function <code><a href='https://rdrr.io/pkg/Rfast/man/Dist.html'>Dist</a></code>; <code>program_distance = "vegan"</code> uses <code><a href='https://rdrr.io/pkg/vegan/man/vegdist.html'>vegdist</a></code>.
See details.</p></td>
    </tr>
    <tr>
      <th>target_space</th>
      <td><p>String.
Default <code>target_space = "multidimensional"</code> applies PERMANOVA and/or PERMDISP to sample-by-sample triangular matrix computed from variable-by-sample data, <code>pc_axes</code> has no effect on computation. <code>target_space = "pca"</code> applies PERMANOVA and/or PERMDISP to sample-by-sample data in PCA space, <code>pc_axes</code> determines number of PCA axes for testing.</p></td>
    </tr>
    <tr>
      <th>pc_axes</th>
      <td><p>Number of PCA axes computed always starting with PCA axis 1.
Default <code>pc_axes = 2</code> computes PCA axes 1 and 2 if <code>target_space = "pca"</code>.
No effect on computation if <code>target_space = "multidimensional"</code>.</p></td>
    </tr>
    <tr>
      <th>pairwise</th>
      <td><p>Logical.
Default <code>pairwise = FALSE</code> computes global test. <code>pairwise = TRUE</code> computes global and pairwise tests.</p></td>
    </tr>
    <tr>
      <th>pairwise_method</th>
      <td><p>String specifying type of correction for multiple testing.
Default <code>"holm"</code>.</p></td>
    </tr>
    <tr>
      <th>permutation_n</th>
      <td><p>Number of permutations resulting in PERMANOVA/PERMDISP test <em>p value</em>.
Default <code>9999</code>.</p></td>
    </tr>
    <tr>
      <th>permutation_seed</th>
      <td><p>Number fixing random generator of permutations.
Default <code>1</code>.</p></td>
    </tr>
    <tr>
      <th>dispersion_type</th>
      <td><p>String indicating quantification of group dispersion whether relative to spatial <code>"median"</code> or <code>"centroid"</code> in PERMDISP.
Default <code>"median"</code>.</p></td>
    </tr>
    <tr>
      <th>samplesize_bias</th>
      <td><p>Logical. <code>samplesize_bias = TRUE</code> for dispersion weighted by number of samples per group in PERMDISP.
Default <code>pairwise = FALSE</code> for no weighting.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a list containing the following elements:</p><ul>
<li><p>pca.snp_loadingsDataframe of principal coefficients of SNPs.
One set of coefficients per PCA axis computed.</p></li>
<li><p>pca.eigenvaluesDataframe of eigenvalues, variance and cumulative variance explained.
One eigenvalue per PCA axis computed.</p></li>
<li><p>pca_sample_coordinatesDataframe showing PCA sample summary. Column <em>Group</em> assigns samples to groups. Column <em>Class</em> specifies if samples "Removed" from PCA or "Projected" onto PCA space.
Sequence of additional columns shows principal components (coordinates) of samples in PCA space (1 column per PCA computed named PC1, PC2, ...).</p></li>
<li><p>test_samplesDataframe showing test sample summary.
Column <em>Group</em> assigns samples to tested groups.
Column <em>Class</em> specifies if samples were used in, or removed from, testing (PERMANOVA and/or PERMDISP).
Column <em>Sample_dispersion</em> shows dispersion of individual samples relative to spatial <code>"median"</code> or <code>"centroid"</code> used in PERMDISP.</p></li>
<li><p>permanova.global_testList showing PERMANOVA table with degrees of freedom, sum of squares, mean sum of squares, <em>F</em> statistic, variance explained (<em>R2</em>) and <em>p</em> value.</p></li>
<li><p>permanova.pairwise_testList showing PERMANOVA table with <em>F</em> statistic, variance explained (<em>R2</em>), <em>p</em> value and corrected <em>p</em> value per pair of groups.</p></li>
<li><p>permdisp.global_testList showing PERMDISP table with degrees of freedoms, sum of squares, mean sum of squares, <em>F</em> statistic and <em>p</em> value.</p></li>
<li><p>permdisp.pairwise_testList showing PERMDISP table with <em>F</em> statistic, <em>p</em> value and corrected <em>p</em> value per pair of groups.
Obtained only if <code>pairwise = TRUE</code>.</p></li>
<li><p>permdisp.biasString indicating if PERMDISP dispersion corrected for number of samples per group.</p></li>
<li><p>permdisp.group_locationDataframe showing coordinates of spatial <code>"median"</code> or <code>"centroid"</code> per group in PERMDISP.</p></li>
<li><p>test.pairwise_correctionString indicating type of correction for multiple testing in PERMANOVA and/or PERMDISP.</p></li>
<li><p>test.permutation_numberNumber of permutations applied to obtain the distribution of <em>F</em> statistic of PERMANOVA and/or PERMDISP.</p></li>
<li><p>test.permutation_seedNumber fixing random generator of permutations of PERMANOVA and/or PERMDISP for reproducibility of results.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>See details in other functions for conceptualization of PCA (<code>smart_pca</code>) (Hotelling 1993), SMARTPCA (Patterson, Price and Reich 2006), PERMANOVA (<code>smart_permanova</code>) (Anderson 2001) and PERMDISP (<code>smart_permdisp</code> (Anderson 2006), types of scaling, ancient projection, and correction for multiple testing.<br /></p>
<p>Users can compute any combination of the three analyses by assigning <code>TRUE</code> or <code>FALSE</code> to <code>pca</code> and/or <code>permanova</code> and/or <code>permdisp</code>.<br /></p>
<p>PERMANOVA and PERMDISP exclude samples (columns) specified in either <code>sample_remove</code> or <code>sample_project</code>.
Projected samples are not used for testing as their PCA coordinates are derived from, and therefore depend on, the coordinates of non-projected samples.<br /></p>
<p>Data read from working directory with SNPs as rows and samples as columns. Two alternative formats: (1) text file of SNPs by samples (file extension and column separators recognized automatically) read using <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>fread</a></code>; or (2) duet of <code>EIGENSTRAT</code> files (see <a href='https://reich.hms.harvard.edu/software'>https://reich.hms.harvard.edu/software</a>) using <code><a href='https://vroom.r-lib.org/reference/vroom_fwf.html'>vroom_fwf</a></code>, including a genotype file of SNPs by samples (<code>*.geno</code>), and a sample file (<code>*.ind</code>) containing three vectors assigning individual samples to unique user-predefined groups (populations), sexes (or other user-defined descriptor) and alphanumeric identifiers.
For <code>EIGENSTRAT</code>, vector <code>sample_group</code> assigns samples to groups retrievable from column 3 of file <code>*.ind</code>.
SNPs with zero variance removed prior to SVD to optimize computation time and avoid undefined values if <code>scaling = "sd"</code> or <code>"drift"</code>.<br /></p>
<p>Users can select subsets of samples or SNPs by introducing a vector including column numbers for samples (<code>sample_remove</code>) and/or row numbers for SNPs (<code>snp_remove</code>) to be removed from computations.
Function stops if the final number of SNPs is 1 or 2.
<code>EIGENSOFT</code> was conceived for the analysis of human genes and its SMARTPCA suite so accepts 22 (autosomal) chromosomes by default.
If &gt;22 chromosomes are provided and the internal parameter <code>numchrom</code> is not set to the target number chromosomes of interest, SMARTPCA automatically subsets chromosomes 1 to 22.
In contrast, <code>smart_mva</code> accepts any number of autosomes with or without the sex chromosomes from an <code>EIGENSTRAT</code> file.<br /></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='smart_pca.html'>smart_pca</a></code>,
<code><a href='smart_permanova.html'>smart_permanova</a></code>,
<code><a href='smart_permdisp.html'>smart_permdisp</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Path to example genotype matrix "dataSNP"</span>
<span class='va'>pathToGenoFile</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"dataSNP"</span>, package <span class='op'>=</span> <span class='st'>"smartsnp"</span><span class='op'>)</span>

<span class='co'># Assign 50 samples to each of two groups and colors</span>
<span class='va'>my_groups</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>as.factor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"A"</span>, <span class='fl'>50</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"B"</span>, <span class='fl'>50</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>; <span class='va'>cols</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"red"</span>, <span class='st'>"blue"</span><span class='op'>)</span>

<span class='co'># Run PCA, PERMANOVA and PERMDISP</span>
<span class='va'>mvaR</span> <span class='op'>&lt;-</span> <span class='fu'>smart_mva</span><span class='op'>(</span>snp_data <span class='op'>=</span> <span class='va'>pathToGenoFile</span>, sample_group <span class='op'>=</span> <span class='va'>my_groups</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Checking argument options selected...</span></div><div class='output co'>#&gt; <span class='message'>Argument options are correct...</span></div><div class='output co'>#&gt; <span class='message'>Loading data...</span></div><div class='output co'>#&gt; <span class='message'>Imported 10000 SNP by 100 sample genotype matrix</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Filtering data...</span></div><div class='output co'>#&gt; <span class='message'>No samples projected after PCA computation</span></div><div class='output co'>#&gt; <span class='message'>10000 SNPs included in PCA &amp; PERMANOVA &amp; PERMDISP computations</span></div><div class='output co'>#&gt; <span class='message'>1 SNPs omitted from PCA computation</span></div><div class='output co'>#&gt; <span class='message'>100 samples included in PCA &amp; PERMANOVA &amp; PERMDISP computations</span></div><div class='output co'>#&gt; <span class='message'>0 samples ommitted from PERMANOVA &amp; PERMDISP analyses</span></div><div class='output co'>#&gt; <span class='message'>Completed data filtering</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scanning for invariant SNPs...</span></div><div class='output co'>#&gt; <span class='message'>Scan complete: no invariant SNPs found</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Checking for missing values...</span></div><div class='output co'>#&gt; <span class='message'>6341 SNPs contain missing values</span></div><div class='output co'>#&gt; <span class='message'>Imputing SNPs with missing values...</span></div><div class='output co'>#&gt; <span class='message'>Imputation with means completed: 10058 missing values imputed</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Scaling values by SNP...</span></div><div class='output co'>#&gt; <span class='message'>Centering and scaling by drift dispersion...</span></div><div class='output co'>#&gt; <span class='message'>Completed scaling using drift</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 0s</span></div><div class='output co'>#&gt; <span class='message'>Computing singular value decomposition using RSpectra...</span></div><div class='output co'>#&gt; <span class='message'>Completed singular value decomposition</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Extracting eigenvalues and eigenvectors...</span></div><div class='output co'>#&gt; <span class='message'>Eigenvalues and eigenvectors extracted</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Completed PCA computation</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Performing PERMANOVA &amp; PERMDISP testing...</span></div><div class='output co'>#&gt; <span class='message'>Construct triangular matrix of sample by sample proximities in multidimensional space...</span></div><div class='output co'>#&gt; <span class='message'>Completed construction of triangular matrix of sample by sample proximities</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 1s</span></div><div class='output co'>#&gt; <span class='message'>Computing variance partioning by PERMANOVA: global test...</span></div><div class='output co'>#&gt; <span class='message'>Completed PERMANOVA: global test</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 4s</span></div><div class='output co'>#&gt; <span class='message'>Compute inter- and intra-group dispersion by PERMDISP: global test...</span></div><div class='output co'>#&gt; <span class='message'>Completed PERMDISP: global test based on group spatial medians</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 4s</span></div><div class='output co'>#&gt; <span class='message'>Completed PERMANOVA &amp; PERMDISP computation</span></div><div class='output co'>#&gt; <span class='message'>Time elapsed: 0h 0m 4s</span></div><div class='output co'>#&gt; <span class='message'>Complete.</span></div><div class='input'><span class='va'>mvaR</span><span class='op'>$</span><span class='va'>pca</span><span class='op'>$</span><span class='va'>pca.eigenvalues</span> <span class='co'># extract PCA eigenvalues</span>
</div><div class='output co'>#&gt;                                      PC1        PC2
#&gt; observed eigenvalues          321.195060 317.773982
#&gt; variance explained              1.209482   1.196600
#&gt; cumulative variance explained   1.209482   2.406083</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>mvaR</span><span class='op'>$</span><span class='va'>pca</span><span class='op'>$</span><span class='va'>pca.snp_loadings</span><span class='op'>)</span> <span class='co'># extract principal coefficients (SNP loadings)</span>
</div><div class='output co'>#&gt;             PC1          PC2
#&gt; 1  0.0046337236 -0.002775738
#&gt; 2  0.0018116572  0.011950491
#&gt; 3  0.0096288656 -0.011907379
#&gt; 4  0.0055371026  0.001682321
#&gt; 5 -0.0046772139  0.002187803
#&gt; 6  0.0004521752  0.002215284</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>mvaR</span><span class='op'>$</span><span class='va'>pca</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>)</span> <span class='co'># extract PCA principal components (sample position in PCA space)</span>
</div><div class='output co'>#&gt;   Group Class        PC1        PC2
#&gt; 1     A   PCA   4.768216  13.413303
#&gt; 2     A   PCA -16.661893  -9.789570
#&gt; 3     A   PCA -25.643492   5.994816
#&gt; 4     A   PCA  -2.453762 -25.043319
#&gt; 5     A   PCA  19.760752   7.214635
#&gt; 6     A   PCA -28.587658   2.656167</div><div class='input'>
<span class='co'># plot PCA</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>mvaR</span><span class='op'>$</span><span class='va'>pca</span><span class='op'>$</span><span class='va'>pca.sample_coordinates</span><span class='op'>[</span>,<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PC1"</span>,<span class='st'>"PC2"</span><span class='op'>)</span><span class='op'>]</span>, cex <span class='op'>=</span> <span class='fl'>2</span>,
     pch <span class='op'>=</span> <span class='fl'>19</span>, col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='va'>my_groups</span><span class='op'>]</span>, main <span class='op'>=</span> <span class='st'>"genotype smartpca"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topleft"</span>, legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>my_groups</span><span class='op'>)</span>, cex <span class='op'>=</span> <span class='fl'>1</span>,
       pch <span class='op'>=</span> <span class='fl'>19</span>, col <span class='op'>=</span> <span class='va'>cols</span>, text.col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>)</span>
</div><div class='img'><img src='smart_mva-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Extract PERMANOVA table</span>
<span class='va'>mvaR</span><span class='op'>$</span><span class='va'>test</span><span class='op'>$</span><span class='va'>permanova.global_test</span>
</div><div class='output co'>#&gt;           Df SumsOfSqs MeanSqs F.Model      R2 Pr(&gt;F)
#&gt; group      1     26183   26183  0.9858 0.00996 0.8463
#&gt; Residuals 98   2602901   26560         0.99004       
#&gt; Total     99   2629084                 1.00000       </div><div class='input'>
<span class='co'># Extract PERMDISP table</span>
<span class='va'>mvaR</span><span class='op'>$</span><span class='va'>test</span><span class='op'>$</span><span class='va'>permdisp.global_test</span> <span class='co'># extract PERMDISP table</span>
</div><div class='output co'>#&gt;           Df    Sum Sq   Mean Sq        F Pr(&gt;F)
#&gt; Groups     1  1.214908 1.2149079 3.387661 0.0705
#&gt; Residuals 98 35.145482 0.3586274       NA     NA</div><div class='input'>
<span class='co'># Extract sample summary and dispersion of individual samples used in PERMDISP</span>
<span class='va'>mvaR</span><span class='op'>$</span><span class='va'>test</span><span class='op'>$</span><span class='va'>test_samples</span>
</div><div class='output co'>#&gt;     Group              Class Sample_dispersion
#&gt; 1       A PERMANOVA/PERMDISP          161.4757
#&gt; 2       A PERMANOVA/PERMDISP          160.9709
#&gt; 3       A PERMANOVA/PERMDISP          161.2118
#&gt; 4       A PERMANOVA/PERMDISP          160.3491
#&gt; 5       A PERMANOVA/PERMDISP          160.4445
#&gt; 6       A PERMANOVA/PERMDISP          162.2998
#&gt; 7       A PERMANOVA/PERMDISP          160.9662
#&gt; 8       A PERMANOVA/PERMDISP          160.9234
#&gt; 9       A PERMANOVA/PERMDISP          161.4040
#&gt; 10      A PERMANOVA/PERMDISP          161.1902
#&gt; 11      A PERMANOVA/PERMDISP          160.2838
#&gt; 12      A PERMANOVA/PERMDISP          161.0910
#&gt; 13      A PERMANOVA/PERMDISP          160.0616
#&gt; 14      A PERMANOVA/PERMDISP          161.9953
#&gt; 15      A PERMANOVA/PERMDISP          161.5489
#&gt; 16      A PERMANOVA/PERMDISP          160.8466
#&gt; 17      A PERMANOVA/PERMDISP          162.0203
#&gt; 18      A PERMANOVA/PERMDISP          161.0723
#&gt; 19      A PERMANOVA/PERMDISP          160.9892
#&gt; 20      A PERMANOVA/PERMDISP          161.2393
#&gt; 21      A PERMANOVA/PERMDISP          160.8561
#&gt; 22      A PERMANOVA/PERMDISP          160.8314
#&gt; 23      A PERMANOVA/PERMDISP          162.2363
#&gt; 24      A PERMANOVA/PERMDISP          160.9487
#&gt; 25      A PERMANOVA/PERMDISP          161.7043
#&gt; 26      A PERMANOVA/PERMDISP          160.0128
#&gt; 27      A PERMANOVA/PERMDISP          161.8192
#&gt; 28      A PERMANOVA/PERMDISP          160.7234
#&gt; 29      A PERMANOVA/PERMDISP          160.9490
#&gt; 30      A PERMANOVA/PERMDISP          160.7225
#&gt; 31      A PERMANOVA/PERMDISP          161.0558
#&gt; 32      A PERMANOVA/PERMDISP          160.7245
#&gt; 33      A PERMANOVA/PERMDISP          161.1502
#&gt; 34      A PERMANOVA/PERMDISP          159.9254
#&gt; 35      A PERMANOVA/PERMDISP          161.1662
#&gt; 36      A PERMANOVA/PERMDISP          161.3988
#&gt; 37      A PERMANOVA/PERMDISP          161.6969
#&gt; 38      A PERMANOVA/PERMDISP          160.5360
#&gt; 39      A PERMANOVA/PERMDISP          161.5771
#&gt; 40      A PERMANOVA/PERMDISP          161.1907
#&gt; 41      A PERMANOVA/PERMDISP          161.8498
#&gt; 42      A PERMANOVA/PERMDISP          161.7793
#&gt; 43      A PERMANOVA/PERMDISP          161.9005
#&gt; 44      A PERMANOVA/PERMDISP          161.7850
#&gt; 45      A PERMANOVA/PERMDISP          161.7027
#&gt; 46      A PERMANOVA/PERMDISP          161.9930
#&gt; 47      A PERMANOVA/PERMDISP          162.2773
#&gt; 48      A PERMANOVA/PERMDISP          160.8876
#&gt; 49      A PERMANOVA/PERMDISP          161.3201
#&gt; 50      A PERMANOVA/PERMDISP          162.0835
#&gt; 51      B PERMANOVA/PERMDISP          161.9058
#&gt; 52      B PERMANOVA/PERMDISP          161.5106
#&gt; 53      B PERMANOVA/PERMDISP          160.9970
#&gt; 54      B PERMANOVA/PERMDISP          160.9550
#&gt; 55      B PERMANOVA/PERMDISP          160.9942
#&gt; 56      B PERMANOVA/PERMDISP          162.4287
#&gt; 57      B PERMANOVA/PERMDISP          161.0528
#&gt; 58      B PERMANOVA/PERMDISP          161.0653
#&gt; 59      B PERMANOVA/PERMDISP          160.5475
#&gt; 60      B PERMANOVA/PERMDISP          161.4787
#&gt; 61      B PERMANOVA/PERMDISP          161.8445
#&gt; 62      B PERMANOVA/PERMDISP          161.1317
#&gt; 63      B PERMANOVA/PERMDISP          161.3229
#&gt; 64      B PERMANOVA/PERMDISP          162.3351
#&gt; 65      B PERMANOVA/PERMDISP          160.9433
#&gt; 66      B PERMANOVA/PERMDISP          161.8845
#&gt; 67      B PERMANOVA/PERMDISP          160.6879
#&gt; 68      B PERMANOVA/PERMDISP          160.8744
#&gt; 69      B PERMANOVA/PERMDISP          163.0843
#&gt; 70      B PERMANOVA/PERMDISP          160.7723
#&gt; 71      B PERMANOVA/PERMDISP          160.9731
#&gt; 72      B PERMANOVA/PERMDISP          161.2441
#&gt; 73      B PERMANOVA/PERMDISP          162.3827
#&gt; 74      B PERMANOVA/PERMDISP          161.2757
#&gt; 75      B PERMANOVA/PERMDISP          161.6916
#&gt; 76      B PERMANOVA/PERMDISP          161.3637
#&gt; 77      B PERMANOVA/PERMDISP          160.7243
#&gt; 78      B PERMANOVA/PERMDISP          162.8925
#&gt; 79      B PERMANOVA/PERMDISP          160.8015
#&gt; 80      B PERMANOVA/PERMDISP          161.7590
#&gt; 81      B PERMANOVA/PERMDISP          161.8004
#&gt; 82      B PERMANOVA/PERMDISP          161.1778
#&gt; 83      B PERMANOVA/PERMDISP          161.3634
#&gt; 84      B PERMANOVA/PERMDISP          161.4223
#&gt; 85      B PERMANOVA/PERMDISP          162.5199
#&gt; 86      B PERMANOVA/PERMDISP          161.3039
#&gt; 87      B PERMANOVA/PERMDISP          161.4728
#&gt; 88      B PERMANOVA/PERMDISP          160.7768
#&gt; 89      B PERMANOVA/PERMDISP          161.7114
#&gt; 90      B PERMANOVA/PERMDISP          162.3565
#&gt; 91      B PERMANOVA/PERMDISP          161.3105
#&gt; 92      B PERMANOVA/PERMDISP          161.0042
#&gt; 93      B PERMANOVA/PERMDISP          161.4582
#&gt; 94      B PERMANOVA/PERMDISP          161.9629
#&gt; 95      B PERMANOVA/PERMDISP          161.2433
#&gt; 96      B PERMANOVA/PERMDISP          161.8619
#&gt; 97      B PERMANOVA/PERMDISP          160.9686
#&gt; 98      B PERMANOVA/PERMDISP          161.7675
#&gt; 99      B PERMANOVA/PERMDISP          161.2074
#&gt; 100     B PERMANOVA/PERMDISP          160.5961</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


