<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using VCF and PLINK formatted files with smartsnp • smartsnp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using VCF and PLINK formatted files with smartsnp">
<meta property="og:description" content="This Vignette shows how you can use VCF and PLINK formatted variant files with smartsnp.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartsnp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Converting_VCF_and_PLINK_formats.html">Using VCF and PLINK formatted files with smartsnp</a>
    </li>
    <li>
      <a href="../articles/aDNA_smartpca_analysis.html">Projecting ancient samples</a>
    </li>
    <li>
      <a href="../articles/mallard_smartpca_analysis.html">Example PCA, PERMANOVA and PERMDISP analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Converting_VCF_and_PLINK_formats_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using VCF and PLINK formatted files with smartsnp</h1>
            
      
      
      <div class="hidden name"><code>Converting_VCF_and_PLINK_formats.Rmd</code></div>

    </div>

    
    
<p>Package <em>smartsnp</em> is not a data-conversion tool. Inspired by the command-line tool SMARTPCA, <em>smartsnp</em> handles SNP datasets in text format and in SMARTPCA formats (uncompressed = EIGENSTRAT or compressed binary = PACKENDANCESTRYMAP) and a general genotype matrix format. However, both VCF (.vcf) and PLINK (.bed) formats are frequently used for storing genetic variation data. In this vignette we provide a quick and robust solution for how to transform these two formats into a general genotype matrix (i.e. where homozygous genotypes are coded as 0 or 2, and heterozygotes as 1) that can be used with the <em>smartsnp</em> package.</p>
<p>The general strategy is to use the <em>plink2</em> software for transforming VCF or PLINK/bed files into a general (transposed) genotype matrix. It is “transposed” because PLINK and VCF files typically have samples in rows, whereas the general input file for <em>smartsnp</em> has samples in columns.</p>
<p>We make heavily use of the plink2 software, which is a comprehensive update to Shaun Purcell’s PLINK original command-line program. Binary downloads and an installation guide is available here: <a href="https://www.cog-genomics.org/plink2" class="uri">https://www.cog-genomics.org/plink2</a></p>
<p>In the PLINK manual, the file format that we transform our data into is called “.traw (variant-major additive component file)”. See here for more information: <a href="https://www.cog-genomics.org/plink/1.9/formats#traw" class="uri">https://www.cog-genomics.org/plink/1.9/formats#traw</a></p>
<p>Note that the <em>.traw</em> format can be directly used with <em>smartsnp</em>, no further data transformation is necessary.</p>
<div id="download-a-small-example-vcf" class="section level2">
<h2 class="hasAnchor">
<a href="#download-a-small-example-vcf" class="anchor"></a>Download a small example VCF</h2>
<p>The R package <em>sim1000G</em> contains a small VCF, an unfiltered region from the 1000 genomes Phase III sequencing data VCF, chromosome 4, CEU samples. We will first load the package and then use this file as an example dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">sim1000G</span><span class="op">)</span>

<span class="co"># First set the current working directory (cwd) to where you want to download the data to (note that the *Downloads* folder might not exist on your computer, e.g. if you have a Windows system):</span>
<span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"~/Downloads/"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">examples_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"examples"</span>, package <span class="op">=</span> <span class="st">"sim1000G"</span><span class="op">)</span>
<span class="va">vcf_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">examples_dir</span>, <span class="st">"region.vcf.gz"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">vcf_file</span>, to <span class="op">=</span> <span class="st">"./"</span><span class="op">)</span> <span class="co"># Copy the file to the cwd</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
</div>
<div id="vcf-to-plink--bed" class="section level2">
<h2 class="hasAnchor">
<a href="#vcf-to-plink--bed" class="anchor"></a>VCF to PLINK (.bed)</h2>
<p>As a first step, we show how to transform a VCF file into a PLINK/bed format. Note that the VCF is gzipped, but <em>plink</em> can directly use gzipped files.</p>
<p>We will use the <em>system</em> function for calling plink. This is effectively the same as running the quoted command on the command line.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"plink --vcf region.vcf.gz --make-bed --out region"</span><span class="op">)</span></code></pre></div>
<p>The –out parameter defines the name of the output file (without any suffix), you can set it to any arbitrary string. After running this command, you will see three files that make up the PLINK/bed format (region.bim, region.bed, region.fam). See the definition of the PLINK/bed (binary biallelic genotype table) file format for more information: <a href="https://www.cog-genomics.org/plink/1.9/formats#bed" class="uri">https://www.cog-genomics.org/plink/1.9/formats#bed</a></p>
<p>The <em>plink</em> software offers a wide range of options for filtering and transforming the data that could be useful for your analysis. See the manual: <a href="https://www.cog-genomics.org/plink/1.9/" class="uri">https://www.cog-genomics.org/plink/1.9/</a></p>
<p>If you don’t want to make use of any further PLINK functionality, then you can also directly transform the VCF file into the .traw format. See section “Directly transforming VCF to raw genotype (.traw)” below.</p>
</div>
<div id="plink-to-raw-genotype--traw" class="section level2">
<h2 class="hasAnchor">
<a href="#plink-to-raw-genotype--traw" class="anchor"></a>PLINK to raw genotype (.traw)</h2>
<p>Now we will use the <em>plink</em> software to transform the .bed file into raw genotypes. Again, note that we will need a “transformed” version since <em>smartsnp</em> assumes that samples are in columns, not rows.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"plink --bfile region --recode A-transpose --out region_genotypeMatrix"</span><span class="op">)</span></code></pre></div>
<p>Again, the –out parameter defines the name of the output file, without the suffix. After running this command, you will see a “region_genotypeMatrix.traw” file. This file can be directly used with smartsnp.</p>
</div>
<div id="vcf-to-raw-genotype--traw" class="section level2">
<h2 class="hasAnchor">
<a href="#vcf-to-raw-genotype--traw" class="anchor"></a>VCF to raw genotype (.traw)</h2>
<p>We could have skipped the intermediate step of transforming the VCF into a PLINK format. The <em>plink</em> software allows to directly transform the VCF into the .traw format.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="st">"plink --vcf region.vcf.gz --recode A-transpose --out region_genotypeMatrix"</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-smartpca" class="section level2">
<h2 class="hasAnchor">
<a href="#running-smartpca" class="anchor"></a>Running smartpca</h2>
<p>The VCF file just contained data from a single group (CEU). However, just to demonstrate that this file can be used with smartsnp we’ll run a simple pca analysis. Importantly, you will have to set the <em>missing_value</em> parameter to “NA”.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First load one of the PLINK files to get the number of samples</span>
<span class="va">numSamples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"region.fam"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># There is just a single group in this data</span>
<span class="va">group_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CEU"</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="va">numSamples</span><span class="op">)</span>

<span class="co"># Running smart_pca</span>
<span class="va">sm.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smart_pca.html">smart_pca</a></span><span class="op">(</span>snp_data <span class="op">=</span> <span class="st">"region_genotype.traw"</span>, 
                    sample_group <span class="op">=</span> <span class="va">group_id</span>,
                    missing_value <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span>

<span class="co"># Here is a plot of the first two components:</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sm.pca</span><span class="op">$</span><span class="va">pca.sample_coordinates</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="figure/data_transformation_example_plot-1.png" title="plot of chunk data_transformation_example_plot" alt="plot of chunk data_transformation_example_plot" style="display: block; margin: auto;"></p>
<p>Voila! Now to go back to the old working directory:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Salvador Herrando-Perez, Christian Huber.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
